---
chapter: 2
lesson: 0
summary: Introduction to Inheritance, Variation, and Genealogy
notebook: None (but maybe recombination/crossing over)
todo: polish
---

# Wright-Fisher 

## Under Construction

## Learning objectives
This chapter ...


## Modeling evolution
The previous chapter introduced trees as a type of data structure,
discussing terminology and methods for using trees to understand
and describe evolutionary history. In this chapter we will turn our
focus to theoretical models underlying the case for treating evolution
as a tree-like process, spanning from the scale of molecules to the
entire tree of life.

Generally, when we think of phylogenetic trees it invokes an image of 
relationships among species, or perhaps even higher-level clades, such as 
taxonomic families or orders. At this scale, the labels at the tips of a
tree, such as *Diptera* or *Arecaceae*, are far-removed from the biological 
sample from which data was collected to infer the tree. In fact, this is 
very often the case. A label at the tip of a tree represents a hypothesis 
for the history of a **lineage** -- a theoretical concept *or model* for 
the idea of a cohesive population that has evolved through time. 
Thus, even when doing something as simple as reading a 
phylogeny, we are inherently making assumptions about a model of 
how populations evolve.
<!-- We can never sample every genome in a species,  -->
<!-- 
## Populations/Species
- Lineages are collections of individuals that ...
- A population exists at any point in time...

## Sampling
- We cannot sample everything, so few samples are used as representatives of lineages.
 -->

A less familiar type of phylogenetic tree is one that represents 
evolutionary relationships among the actual units of heritable 
information that descend from one generation to the next. At this scale
the tips of a phylogenetic tree *are in fact the biological samples for 
which the relationships are being depicted*. An example is a phylogeny
of *gene copies*, also called a *genealogy*. 

<!-- SOME tree-thinking type image here, of gene copies, to pops, to species -->


## Gene copies 
As a diploid organism your genome exists in two copies, one inherited from
each of your two parents. These two copies are likely not identical, having 
accumulated mutations over time. Here the term *copy* is used somewhat 
confusingly to refer to the fact that these units can replicate themselves
and be inherited from one generation to the next. It is not a reference
to whether or not they share mutations. In a population of size N there will
exist 2N gene copies for any given gene. Here, similarly, the term *gene* 
is also somewhat misleading, since we do not actually care in this context
whether or not it is a coding or non-coding region of the genome. A gene 
copy can simply be thought of as a haploid region of a genome.


## Genealogies
A genealogy is a record of ancestor-descendant relationships. It is 
similar to a phylogenetic tree, and often modeled as one, but in reality
can include complex network-like relationships as well. Most people
are familiar with the concept of genealogy as it relates to tracing 
family trees. For example, a genealogy of your family would show that 
you are descended from your two parents (one generation back), and that
they are descended from your four grandparents (two generations back), 
and on and on. Your siblings, cousins, uncles/aunts, and other relatives
could be represented as well. This type of genealogy is a *genealogy of
individuals*, and although it can be similar to a *genealogy of gene copies*
at very short time scales, these two trees deviate significantly over time.
... The difference between a genetic and individual-level genealogy is 
detailed in a later chapter.

<!-- ==image of individual genealogy== -->

## Genetic genealogy
A genetic genealogy, in contrast to an individual-level genealogy, is a
genealogy of gene copies for some region of a genome. Although there exists
only one genealogy for you as an individual, your genome is composed of 
millions of distinct genealogies of gene copies. This is because each 
region of your genome can have a different genealogy from other regions, 
and even within a region, your two gene copies will trace back different 
histories. Each chromosome in an individual's genome is a *mosaic* of 
recombined segments from its many different ancestors.

For example, the two gene copies at gene A in an individual's genome may 
come, respectively, from its father's father, and mother's father (i.e.,
the individual's two grandfathers). At gene B, by contrast, the two copies 
came from its father's mother and mother's mother, respectively. Thus, 
no information from either of the individual's grandmothers is present at 
gene A, and no information from either of its grandfathers is present at 
gene B. *The gene copies at these two regions have different genealogical 
histories.*


<figure markdown>
  ![WF-plot-2.2.3](https://raw.githubusercontent.com/eaton-lab/eaton-lab.github.io/master/slides/fundamentals/8/genealogies-1.svg){width="400", loading=lazy}
  <figcaption>Fig. 2.0.0 Ancestry of two gene copies at a single locus.</figcaption>
</figure>


This process continues backwards in time. The gene copies that exist
in each grandparent were also inherited from only a fraction of that 
grandparent's ancestors. In fact, the probability that one of an individual's 
ancestors continues to be one of its *genetic ancestors* -- by contributing at 
least some genetic material to its genome -- decreases rapidly with the 
number of generations back to that ancestor. ==References to some interesting 
papers, etc. on this topic.==
<!-- We will return to a statistical view of this rate in a later lesson. -->
<!-- We will return to the topic of recombination in detail in a later chapter.  -->
<!-- Aside: This leads to an interesting pattern in which an individual genealogical -->
<!-- history may deviate significantly from their genetic history... -->
<!-- ==image of haplotype genealogy== -->


## Independent assortment
Even if we make the simplifiying assumption that recombination is absent, 
different chromosomes in a genome are still expected to have different 
genealogical histories, since each is inherited independently of the other.
For example, consider a diploid organism with two copies of chromosomes A 
and B, labeled (A1, A2, B1, and B2). When this organism forms haploid gametes
during reproduction, there is equal probability that a gamete will inherit
either copy of either chromosome, such that A1B1 is just as likely as 
A1B2, A2B1, or A2B2. This is called the law independent assortment.

Remember Mendel's pea experiments? Well, that's where this idea comes from.
The rule of *segregation and independent assortment* was not only an 
interesting experiment for understanding overdominance in the phenotypes of 
wrinkled peas, but rather, the fundamental finding of this work was that 
*genes get passed on to descendants as discrete units independently of one 
another*. This process turns out to be relevant to modeling genealogical 
relationships of gene copies, where the genealogy of gene copies at one 
gene in a genome can be completely independent of the genealogy for gene 
copies at different genes in the same diploid organism's genome. The 
distribution of different genealogies throughout a single genome, and among populations, is the focus of lessons in this chapter. 

<!-- 
In the absence of recombination a chromosome can represent a single 
*gene copy* -- a genetic unit that is inherited independantly of other 
genetic units (other chromosomes copies) by its offspring. When 
recombination does occur, however, a chromosome is subdivided by crossover 
events, such that smaller blocks of one genome are recombined with another, 
in which case each non-recombined block of the chromosome represents a 
gene copy. Perhaps the most fitting description of a
gene copy is something for which multiple copies can trace back a 
genealogy without recombination...? -->


## Nested hierarchy
- A genealogy represents a special type of phylogenetic tree.
- Evolutionary relationships exist in a nested hierarchy. 
	- haplotype genealogy nested within individual genealogies
	- ind genealogies nested within population/species trees.
- Higher-level phylogenetic trees are a model for the expected 
distribution of genealogical trees...


## Effective population size
One of the most challenging concepts in population genetics is that of 
*effective population size* ($N_e$). This is not to be confused with
the population census size, which is the number of individuals in a 
population. 

The concept of effective population size was first developed by Sewall 
Wright (1931, 1933) as a metric for describing the rate of genetic drift,
or inbreeding. Specifically, it was developed in reference to populations
evolving under a WF process, where an idealized population will 
experience changes in allele frequencies over time due to only a single
evolutionary force, genetic drift caused by a finite population size ($N_e$). 

The effective population is usually smaller than the actual population 
census size, representing the effects of deviations from the assumptions 
of an idealized population. In the WF process model, it is easy to 
interpret $N_e$, since it is equivalent to the census population size, 
N. However, in more complex models, and especially in real populations, 
$N_e$ becomes a summary statistic for the impacts of many possible 
evolutionary processes that violate the assumptions of an idealized 
population, causing genetic drift and inbreeding at higher rates
than would occur in a population of the same size if it were an idealized
population. 

For example, an idealized population of census size N=10,000 will have 
an effective population size of $N_e$=10,000. By contrast, a different 
population of census size N=10,000 that experiences non-random mating
may have an effective population size of $N_e$=6,000. This value can be
interpreted to mean that the second population will experience genetic
drift at the same rate as an idealized population of size N=6,000, even
though it in fact contains 10,000 individuals.

<!-- 
Because in the standard WF process the only way in which it differs from
an idealized population is by having a finite population size, the WF process
provides a convenient way to study the effects of $N_e$, since in this case, 
and this case alone, $N_e$ will be equal to the population census size. 
Thus, we can examine how raising or lowering $N_e$ affects the rate of 
fixation of alleles.  -->

[Review papers] continue to be written on the subject of $N_e$, and how it 
should be interpreted, and the development of methods for measuring $N_e$ 
as a descriptor of evolution remain at the forefront of population genetics.
Recent work includes methods for examining $N_e$ as a variable metric 
spatially across regions of genomes in a population, as well as for 
estimating changes in $N_e$ through time. As we will see, there are 
also alternative statistical approaches for describing $N_e$ based on 
probabilities of coalescence, which we will discuss in the next lesson.

[Review papers]: https://www.nature.com/articles/hdy201643. 
## Inference and Truth
- A true genealogical history exists, but cannot be perfectly inferred.
- ...

## Summary
- This was an introduction to what is to come in this chapter.
- ...
