---
chapter: 2
lesson: 3
title: Coalescent Trees
summary: WF -> genealogy -> and genetic drift.
notebook: docs/chapter-1/notebooks/chapter-11-2.ipynb
---


# Coalescent trees

## Under Construction

## Learning objectives

By the end of this lesson you will be familiar with:

1. Coalescent theory.
2. Modeling waiting times between coalescent events as a probability distribution.
3. Generating coalescent trees.
4. Coalescent trees as random variables (model assumptions).


This lesson will introduce Coalescent theory and the process by which 
coalescent trees can be randomly generated, and treated as stochastic
random variables. For this, we will use the `numpy` library to access
statistic distributions, and `toytree` to construct and visualize trees.
By the end of the lesson you will write your own simple coalescent 
simulator to understand the basic concepts and steps underlying this 
process.

```python
import numpy as np
import toytree
```

## Wright-Fisher models and coalescence
In session 2.2 we introduced Wright-Fisher (WF) models to simulate evolution
in a discrete random mating population of size N diploid gene copies. Over
multiple generations of random mating, genetic drift arises from random 
variation in which gene copies are replicated from one generation to the next.
This has the consequence that genetic variation in the population is slowly 
reduced over time, as different alleles become lost or fixed. 
A second important outcome of this process is that it creates a 
network of ancestor-descendant relationships among gene copies (a genealogy). 
The loss of genetic variation over time, and the genealogical relationship
among gene copies, are inherently linked. If we run a simulation long enough
for one allele to fix, and then construct a genealogy of gene copies in 
the current population, we will find that they all trace back to an 
ancestral gene copy that had the same allelic state. 
The allelic state of that ancestor was inherited by all of the present 
gene copies (for now we are ignoring the possibility of new mutations
generating new alleles). 

!!! tip "A genealogy is informative about the amount of genetic variation in the population."
    
    If we run a WF simulation long enough for one allele to fix (e.g., green), 
    and then construct a genealogy of gene copies in the current population, 
    we will find that they all trace back to one or more ancestral gene 
    copies that had the same allelic state. This allele was inherited by 
    all of the present gene copies from their ancestors. The random genealogy
    connecting these gene copies shows the history of these copies of that 
    allele. (For now we are ignoring the possibility of new mutations
    generating new alleles). 

```python
wf = toytree.learn.wright_fisher_simulation(...)
```
<!-- <figure markdown>
  ![WF-plot-2.2.3](../chapter-1/notebooks/fig.2.2.3.svg){width="700", loading=lazy}
  <figcaption>Fig. 2.2.3 Coalescence in a WF process.</figcaption>
</figure> -->


## Coalescence of two samples
Sewall Wright (1931) was the first to describe the process of coalescence that
arises within a WF process. Every pair of gene copies eventually traces
back to a common ancestor, a process referred to as **coalescence**. 
The expected time until this happens can be described using probability
theory, and turns out to rely entirely on a single population parameter:
the effective populations size, Ne. 

The probability that two gene copies share the same ancestor one generation
in the past is 1 / 2Ne. This may look familiar: In a previous lesson we 
described that the strength of genetic drift is 1 / 2Ne. The fact that 
both of these processes occur at the same rate suggests the connection 
between them. 
You can compute this by thinking of each of the two gene copies in turn. 
The first gene copy must have an ancestor in the previous
generation, so the probability of this event is 1. The next gene copy may or
may not have the same ancestor in the previous generation. The probability
that it does is 1/2N, because there are 2N possible gene copies to choose
from. 
<!-- This probability was first derived by Sewall Wright (1931), and forms the basis on which many statistical findings were derived.  -->

We can describe this process as the outcome of a [Bernoulli trial], 
a random variable that returns True with probability p, and 
False with probability 1 - p. This type of probability distribution is 
common in statistics. Often, we are interested in the question of *how
many trials will need to occur, on average, until we observe a success?*
This question can also be described by a probability distribution, called
the [geometric distribution]. The mean expectation of the geometric 
distribution for the time until a successful event is observed, given its
probability $p$, is $E[t] = 1 / p$. 
To translate this back into our realm of interest: the probability that 
two samples coalesce one generation ago is $p = 1 / 2Ne$. Thus, the expected
number of generations until two samples coalesce is $E[t] = 2Ne$. 


### TODO

- a function to add mutations to the tree at rate u.
- the genetic diversity in a population theta = 4Ne * u
- estimate Ne based on genetic diversity


<!-- - Bernoulli trial -->
<!-- - Geometric distribution -->
<!-- - Exponential waiting time -->

<!-- https://vitalflux.com/geometric-distribution-explained-with-python-examples/ -->

<!-- Geometric Distribution is 
 -->
<!-- At each successive preceding generation, the probability of coalescence is geometrically distributed—that is, it is the probability of noncoalescence at the t − 1 preceding generations multiplied by the probability of coalescence at the generation of interest: 
 -->



## Kingman coalescent
J. F. C. Kingman was a mathematician (1982) that used advances in probability
theory to develop a generalization of the expectations we can derive
for two sampled gene copies, and instead make claims about any arbitrarily 
large number of sampled gene copies, which we will call $k$. 

### Derivation
Let's start by asking, what is the probability that two samples do not 
coalesce in the previous generation -- that they don't share a common 
ancestor one generation back. Viewed backwards in time, the first gene copy 
samples an ancestor, and the second gene copy has a 1 - 1/2N probability of 
sampling a different ancestor. We can then extend this to ask what is the
probability that a third sample does not coalesce with either of those two
one generation back, which 1 - 2 / 2N. We can repeat this process for 
all $k$ sampled gene copies of interest to get a series that represents
the probability that $k$ samples do not coalesce one generation ago:

$$ P(\text{no coal}| k, N) = (1 - \frac{1}{2N}) \times (1 - \frac{2}{2N}) \times (1 - \frac{3}{2N}) \times ... (1 - \frac{k}{2N}) $$

The right side of this equation can be multiplied out to simplify its
representation as:

$$ P(\text{no coal} | k, N) = 1 - [1 + 2 + 3 + ... (k - 1)] \times \frac{1}{2N} + (\text{terms in} \frac{1}{N^2}) $$

We will now use two tricks to simplify this. First, the sum of integers 
from 1 to k - 1 can be written as k(k-1)/2. Second, we will simply 
discard the term on the right, which corresponds to relatively rare 
instances where more than two sampled gene copies would coalesce 
in the same generation (more on this below). And so we get:

$$ P(\text{no coal} | k, N) = 1 - \frac{k(k-1)}{2} \times \frac{1}{2N} $$

which finally simplifies to the more commonly represented form:

$$ P(\text{no coal} | k, N) = 1 - \frac{k(k-1)}{4N} $$

Finally, if we describe P(coal | k, N) as 1 - P(\text{no coal}| k, N) we
get a solution for the probability that *one* coalescent event occurs in
the previous generation given $k$ sampled gene copies in a population with
effective size N. 

$$ P(\text{coal} | k, N) = \frac{k(k-1)}{4N} $$

Kingman's coalescent is actually an *approximation* of the coalescent, 
because we next simply ignore the second bit of the equation above. Kingman (1982) 
demonstrated that this was a pretty good approximation as long as k(k-1)
is much smaller than the population size N. By throwing away the second
part of this equation we are discarding the probability that two or more
coalescent events among our sampled gene copies occur during the 
same generation. This is thus an assumption of the n-coalescent: 
generations are discrete and non-overlapping, and only one coalescent event
occurs per generation among a set of observed samples.

### Statistical Distribution
This probability can be thought of like a biased coin flip,  
where a heads indicates a coalescent event, and a tails means no
coalescent event. It is generally much more common that there will be no 
coalescent event than there being a coalescence. We can repeat this many
times to ask how many trials will be required, on average, until we observe
a heads. This type of question belongs to the realm of the 
[geometric distribution](https://en.wikipedia.org/wiki/Geometric_distribution):
The probability distribution of the number X of Bernoulli trials needed 
to get one success, e.g., { 1, 2, 3, ...}. 

Given the probability of an event, p, the mean number of trials before a 
success is observed under the geometric probability distribution is 1/p. 
We can translate this into our scenario. We have a statement for the 
probability of a coalescent event $P(\text{coal} | k, N)$, where each 
generation represents a trial where coalescence could occur. And so the mean 
(Expectation) for the time until the first observed coalescence event 
(which we will call $t_k$) is 1 over this probability:

$$ E[t_k] = \frac{4N}{k(k-1)} $$

This expectation of the coalescent model seems relatively simple, but 
it turns that it can be used in a lot of powerful ways.

## Genealogies as random variables

Waiting time until coalescent, but who will coalesce? Hopefully you guessed
it: a random pair out of the k gene copies. This is because we are still
working within the assumptions of the Wright-Fisher process and the model
of an idealized population with discrete non-overlapping generations and 
random mating. 


We can simulate this process by randomly sampling gene copies to 
serve as parents of gene copies in successive generations.
This process can be repeated many times, from different starting seeds, 
and we would likely get a different outcome nearly every time, representing
*a random sample over the possible ways in which some number of gene 
copies could be related.* This turns out 

If we think of these genealogies as trees, rather than just a collection
of waiting times. 

... It can often be difficult at first to understand how a collection of
waiting times can represent a genealogy (tree). How does this process 
group samples together to form a tree from the waiting times? Randomly. 
Remember, random mating is fundamental to our model assumptions.
But how could a random tree possibly be useful compared to real data, 
where some samples will surely be more closely related than others most
of the time? ...


<!-- show several random 10 tip genealogies -->


<!-- show how sampling 10 out of 50 captures all deep splits (usually)-->


<!-- 
## Genetic variation on genealogies
Take note, this is not the probability that they
share a specific common ancestor in the previous generation, but rather 
a probability of whether or not they 
 -->

## Practical applications
Rather than keep track of every single gene 
copy in a large population, we can instead formulate expectations for 
how a small number of samples are expected to be related. It thus 
serves a practical purpose for both simulations and empirical studies
of populations.


## Assumptions of the coalescent
Kingman's n-coalescent is by far the most widely used model of its kind,
but several variants also exist which make slightly different assumptions
that may be more appropriate for some study systems. 

- k is much much smaller than Ne.
- non-overlapping generations.
- ...
