
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      <link rel="icon" href="/images/favicon.ico">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-9.0.11">
    
    
      
        <title>1 amaranthus dioecy - Phylogenetic Data Science (IN DEVELOPMENT)</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.0d440cfe.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.2505c338.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  


  <script id="__analytics">function __md_analytics(){function n(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],n("js",new Date),n("config","G-HV34HZZD5N"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){this.value&&n("event","search",{search_term:this.value})}),document$.subscribe(function(){var a=document.forms.feedback;if(void 0!==a)for(var e of a.querySelectorAll("[type=submit]"))e.addEventListener("click",function(e){e.preventDefault();var t=document.location.pathname,e=this.getAttribute("data-md-value");n("event","feedback",{page:t,data:e}),a.firstElementChild.disabled=!0;e=a.querySelector(".md-feedback__note [data-md-value='"+e+"']");e&&(e.hidden=!1)}),a.hidden=!1}),location$.subscribe(function(e){n("config","G-HV34HZZD5N",{page_path:e.pathname})})});var e=document.createElement("script");e.async=!0,e.src="https://www.googletagmanager.com/gtag/js?id=G-HV34HZZD5N",document.getElementById("__analytics").insertAdjacentElement("afterEnd",e)}</script>

  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#workflow-diagram" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Phylogenetic Data Science (IN DEVELOPMENT)" class="md-header__button md-logo" aria-label="Phylogenetic Data Science (IN DEVELOPMENT)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Phylogenetic Data Science (IN DEVELOPMENT)
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              1 amaranthus dioecy
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent=""  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 0-7 7c0 2.38 1.19 4.47 3 5.74V17a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1v-2.26c1.81-1.27 3-3.36 3-5.74a7 7 0 0 0-7-7M9 21a1 1 0 0 0 1 1h4a1 1 0 0 0 1-1v-1H9v1Z"/></svg>
            </label>
          
        
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 1 7 7c0 2.38-1.19 4.47-3 5.74V17a1 1 0 0 1-1 1H9a1 1 0 0 1-1-1v-2.26C6.19 13.47 5 11.38 5 9a7 7 0 0 1 7-7M9 21v-1h6v1a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1m3-17a5 5 0 0 0-5 5c0 2.05 1.23 3.81 3 4.58V16h4v-2.42c1.77-.77 3-2.53 3-4.58a5 5 0 0 0-5-5Z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/eaton-lab/phylogenetic-data-science" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Phylogenetic Data Science (IN DEVELOPMENT)" class="md-nav__button md-logo" aria-label="Phylogenetic Data Science (IN DEVELOPMENT)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Phylogenetic Data Science (IN DEVELOPMENT)
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/eaton-lab/phylogenetic-data-science" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        About/Ethos
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../installation/" class="md-nav__link">
        Installation/Software
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../course/" class="md-nav__link">
        Columbia course
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
      
      
      
        <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
          0 - Coding Bootcamp
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          0 - Coding Bootcamp
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/getting-started/" class="md-nav__link">
        Getting started
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.0-installation-conda/" class="md-nav__link">
        0.0 - Installation (conda)
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.1-jupyter-and-binder/" class="md-nav__link">
        0.1 - Python Jupyter/Binder
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.2-python-types/" class="md-nav__link">
        0.2 - Python types
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.3-python-funcs/" class="md-nav__link">
        0.3 - Python functions
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.4-python-classes/" class="md-nav__link">
        0.4 - Python classes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.5-python-style/" class="md-nav__link">
        0.5 - Python style
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
      
      
      
        <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
          1 - Trees as data
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          1 - Trees as data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.1-introduction/" class="md-nav__link">
        x.1 - Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.2-tree-thinking/" class="md-nav__link">
        x.2 - Tree Thinking
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.3-visualization/" class="md-nav__link">
        x.3 - Visualization
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.4-features-of-trees/" class="md-nav__link">
        1.4 - Tree data structure
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.5-code-node-class/" class="md-nav__link">
        1.5 - Code - Node class
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.6-code-tree-traversal/" class="md-nav__link">
        1.6 - Code - Tree traversal
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.7-code-tree-class/" class="md-nav__link">
        1.7 - Code - Tree class
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.6-exercises.md" class="md-nav__link">
        x.7 - Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
      
      
      
        <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
          2 - Evolutionary models
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          2 - Evolutionary models
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.0-modeling-evolution/" class="md-nav__link">
        2.0 - Modeling evolution
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.1-idealized-population/" class="md-nav__link">
        2.1 - Idealized population
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.2-wright-fisher/" class="md-nav__link">
        2.2 - Wright-Fisher Process
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.3-coalescent/" class="md-nav__link">
        2.3 - The Coalescent
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.4-coalescent-simulation/" class="md-nav__link">
        2.4 - Coalescent simulation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.x-exercises/" class="md-nav__link">
        2.x - Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
      
      
      
        <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
          3 - Mutation Models
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_7">
          <span class="md-nav__icon md-icon"></span>
          3 - Mutation Models
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-3/3.x-DNA-Markov.md" class="md-nav__link">
        3.0 - Modeling Evolution
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        3.x - Probabilistic and Generative Models
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-3/3.2-markov-models/" class="md-nav__link">
        3.2 - Markov models
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../3.x-other-models.md" class="md-nav__link">
        3.x - Other models
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../1-reading-list.md" class="md-nav__link">
        Reading and Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
      
      
      
        <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
          4 - Data
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_8">
          <span class="md-nav__icon md-icon"></span>
          4 - Data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../4.1-data-formats.md" class="md-nav__link">
        x.1 - data formats
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../4.2-strings-and-matrices.md" class="md-nav__link">
        x.2 - strings, tables, and matrices
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.x-simulating-traits.md" class="md-nav__link">
        x.x - Code - simulating data/traits
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.x-sequence-alignment.md" class="md-nav__link">
        x.x - Code - sequence alignment
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.x-converting-file-formats.md" class="md-nav__link">
        x.x - Code - converting file formats
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../read, write, parse, find, and clean phylip, fasta, fastq, and vcf data." class="md-nav__link">
        Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
      
      
      
        <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
          5 - Phylogenetic Inference
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_9">
          <span class="md-nav__icon md-icon"></span>
          5 - Phylogenetic Inference
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.0 - introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.1 - history
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.2 - choosing a method
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.3 - ..." class="md-nav__link">
        None
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.4 - ..." class="md-nav__link">
        None
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_10" >
      
      
      
        <label class="md-nav__link" for="__nav_10" id="__nav_10_label" tabindex="0">
          6 - Inference (Parsimony)
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_10_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_10">
          <span class="md-nav__icon md-icon"></span>
          6 - Inference (Parsimony)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-6/6.0-fitch-parsimony/" class="md-nav__link">
        6.0 - Fitch Parsimony
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-6/6.1-fitch-multiple-traits/" class="md-nav__link">
        6.1 - Character matrices
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        6.2 - Sankoff Parsimony
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_11" >
      
      
      
        <label class="md-nav__link" for="__nav_11" id="__nav_11_label" tabindex="0">
          7 - Inference (Distance-based)
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_11_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_11">
          <span class="md-nav__icon md-icon"></span>
          7 - Inference (Distance-based)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.0-introduction/" class="md-nav__link">
        7.0 - Distance based trees
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.1-least-squares/" class="md-nav__link">
        7.1 - Code - Least-squares
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.2-upgma/" class="md-nav__link">
        7.2 - Code - UPGMA
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.3-neighbor-joining/" class="md-nav__link">
        7.3 - Code - Neighbor Joining
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.4-applied/" class="md-nav__link">
        7.4 - Applied examples
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.5-new-applications/" class="md-nav__link">
        7.5 - New Applications
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_12" >
      
      
      
        <label class="md-nav__link" for="__nav_12" id="__nav_12_label" tabindex="0">
          8 - Inference (Likelihood)
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_12_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_12">
          <span class="md-nav__icon md-icon"></span>
          8 - Inference (Likelihood)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.0 - Maximum Likelihood
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.1 - Genealogies
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.2 - Sequences
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.3 - Inferring
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Reading and Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_13" >
      
      
      
        <label class="md-nav__link" for="__nav_13" id="__nav_13_label" tabindex="0">
          9 - Multispecies Coalescent
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_13_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_13">
          <span class="md-nav__icon md-icon"></span>
          9 - Multispecies Coalescent
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.0-intro.md" class="md-nav__link">
        9.0 - Species trees
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.1-likelihood.md" class="md-nav__link">
        9.1 - Coalescent likelihood
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.2-genealogical.md" class="md-nav__link">
        9.2 - Genealogical thinking
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.2-genetics.md" class="md-nav__link">
        9.3 - Genetic thinking
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.3-MSC.md" class="md-nav__link">
        9.4 - MSC likelihood
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Reading and Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#workflow-diagram" class="md-nav__link">
    Workflow diagram
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#study-description" class="md-nav__link">
    Study description
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#fetch-fastq-data" class="md-nav__link">
    Fetch fastq data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kmerkit-analysis" class="md-nav__link">
    Kmerkit analysis
  </a>
  
    <nav class="md-nav" aria-label="Kmerkit analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#initialize-a-new-project" class="md-nav__link">
    Initialize a new project
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#read-trimming-optional" class="md-nav__link">
    Read trimming (optional)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#count-kmers" class="md-nav__link">
    Count kmers
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter-kmers" class="md-nav__link">
    Filter kmers
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#extract-reads" class="md-nav__link">
    Extract reads
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#assemble-contigs" class="md-nav__link">
    Assemble contigs
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reproducibility" class="md-nav__link">
    Reproducibility
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#todo-post-pipeline-analysis-api" class="md-nav__link">
    TODO: Post-pipeline analysis API
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  



  <h1>1 amaranthus dioecy</h1>

<h2 id="workflow-diagram">Workflow diagram<a class="headerlink" href="#workflow-diagram" title="Permanent link">&para;</a></h2>
<div class="highlight"><pre><span></span><code>%%{init: {&#39;theme&#39;: &#39;dark&#39;, &quot;flowchart&quot; : { &quot;curve&quot; : &quot;basis&quot; } } }%%
graph LR
    0(kinit)
    1(ktrim)
    A(kcount)
    B(kfilter)
    C(kextract)
    D(kassemble)
    0 --&gt; 1 --&gt; A --&gt; B --&gt; C --&gt; D

linkStyle default stroke-width:2px,fill:none,stroke:grey;       
</code></pre></div>
<h2 id="study-description">Study description<a class="headerlink" href="#study-description" title="Permanent link">&para;</a></h2>
<p>Here we re-implement the study by Neves et al. to detect a male linked 
genomic region involved in sex determination in the dioecious plant species 
<em>Amaranthus palmeri</em>. This study uses pool-seq to sequence four populations 
composing male and female plants from two geographically distinct populations.</p>
<p><cite>Cátia José Neves, Maor Matzrafi, Meik Thiele, Anne Lorant, Mohsen B 
    Mesgaran, Markus G Stetter, Male Linked Genomic Region Determines Sex in 
    Dioecious Amaranthus palmeri, Journal of Heredity, Volume 111, Issue 7, 
    October 2020, Pages 606–612, 
    <a href=https://doi.org/10.1093/jhered/esaa047>https://doi.org/10.1093/jhered/esaa047</a></cite></p>
<h2 id="fetch-fastq-data">Fetch fastq data<a class="headerlink" href="#fetch-fastq-data" title="Permanent link">&para;</a></h2>
<p>If you wish to follow along you can dowload the data from ERA with the
instructions below: </p>
<ul>
<li>ERR4161581 (<em>California_male_pool</em>),</li>
<li>ERR4161582 (<em>California_female_pool</em>),</li>
<li>ERR4161583 (<em>Kansas_male_pool</em>),</li>
<li>ERR4161584 (<em>Kansas_female_pool</em>)</li>
</ul>
<details class="abstract">
<summary>Download fastq data using sra-tools</summary>
<p>Download the latest version of the sratools from <a href="https://github.com/ncbi/sra-tools/wiki/01.-
Downloading-SRA-Toolkit">https://github.com/ncbi/sra-tools/wiki/01.-Downloading-SRA-Toolkit</a> by selecting the compiled binaries that are 
appropriate for your system (e.g., Linux or MacOSX). (I really do 
recommend that you use the latest version since this software is updated
frequently and does not maintain compatibility with older versions.)
Run the command below to download the fastq data for this study 
into a new directory. The total file size will be about 140Gb.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># make a directory to store the raw fastq data files</span>
mkdir<span class="w"> </span>-p<span class="w"> </span>./fastq-data

<span class="c1"># download files to the specified fastq directory</span>
<span class="k">for</span><span class="w"> </span>run<span class="w"> </span><span class="k">in</span><span class="w"> </span>ERR4161581<span class="w"> </span>ERR4161582<span class="w"> </span>ERR4161583<span class="w"> </span>ERR4161584<span class="p">;</span><span class="w"> </span><span class="k">do</span>
<span class="w">    </span>fasterq-dump<span class="w"> </span>--progress<span class="w"> </span>--outdir<span class="w"> </span>./fastq-data<span class="w"> </span>--temp<span class="w"> </span>.<span class="w"> </span><span class="nv">$run</span><span class="p">;</span>
<span class="k">done</span>
</code></pre></div>
</details>
<h2 id="kmerkit-analysis">Kmerkit analysis<a class="headerlink" href="#kmerkit-analysis" title="Permanent link">&para;</a></h2>
<h3 id="initialize-a-new-project">Initialize a new project<a class="headerlink" href="#initialize-a-new-project" title="Permanent link">&para;</a></h3>
<p>Start a new project by entering a name and working directory to <code>init</code>, 
representing the filename prefix and location where all files will be saved.
The directory will be created if it doesn't yet exist. Multiple input fastq 
files can be entered as arguments, or multiple can be selected using a 
regular expression selector like below. Paired reads are automatically 
detected based on name matching. Sample names are automatically extracted 
from file names. Paired file names should be identical up until some 
delimiter (e.g., <code>_</code>), after which the tail is trimmed (<code>_R1.fastq.gz</code>). 
You can set the delimiter explicitly.</p>
<div class="highlight"><pre><span></span><code>kmerkit<span class="w"> </span>init<span class="w"> </span>--name<span class="w"> </span>dioecy<span class="w"> </span>--workdir<span class="w"> </span>/tmp<span class="w"> </span>--delim<span class="w"> </span><span class="s2">&quot;_&quot;</span><span class="w"> </span>./fastq-data/*.gz<span class="w"> </span>
</code></pre></div>
<p>This step creates a project JSON file, which will contain fully reproducible 
information about each step in a kmerkit analysis. It will be updated 
upon each kmerkit module that is run. This file can be read directly, 
or, you can view it easily by calling <code>kmerkit stats</code>, like below.</p>
<details class="tip">
<summary>kmerkit stats output</summary>
<div class="highlight"><pre><span></span><code>kmerkit<span class="w"> </span>stats<span class="w"> </span>--json<span class="w"> </span>/tmp/dioecy.json
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="err">Projec</span><span class="kc">t</span><span class="w"> </span><span class="err">JSON</span><span class="w"> </span><span class="err">da</span><span class="kc">ta</span><span class="p">:</span>
<span class="p">{</span>
<span class="w">    </span><span class="nt">&quot;name&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;Neves&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;workdir&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/pinky/deren/palmeri&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;versions&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="nt">&quot;kmerkit&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;0.0.12&quot;</span><span class="p">,</span>
<span class="w">        </span><span class="nt">&quot;kmc&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;3.1.1&quot;</span><span class="p">,</span>
<span class="w">        </span><span class="nt">&quot;gemma&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;0.9.83&quot;</span><span class="p">,</span>
<span class="w">        </span><span class="nt">&quot;sklearn&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;0.24.1&quot;</span>
<span class="w">    </span><span class="p">},</span>
<span class="w">    </span><span class="nt">&quot;kinit&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="nt">&quot;data&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">            </span><span class="nt">&quot;ERR4161581&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">[</span>
<span class="w">                </span><span class="s2">&quot;/pinky/DATASTORE/Amaranthus_palmeri_PRJEB38372/ERR4161581_1.fastq.gz&quot;</span><span class="p">,</span>
<span class="w">                </span><span class="s2">&quot;/pinky/DATASTORE/Amaranthus_palmeri_PRJEB38372/ERR4161581_2.fastq.gz&quot;</span>
<span class="w">            </span><span class="p">],</span>
<span class="w">            </span><span class="nt">&quot;ERR4161582&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">[</span>
<span class="w">                </span><span class="s2">&quot;/pinky/DATASTORE/Amaranthus_palmeri_PRJEB38372/ERR4161582_1.fastq.gz&quot;</span><span class="p">,</span>
<span class="w">                </span><span class="s2">&quot;/pinky/DATASTORE/Amaranthus_palmeri_PRJEB38372/ERR4161582_2.fastq.gz&quot;</span>
<span class="w">            </span><span class="p">],</span>
<span class="w">            </span><span class="nt">&quot;ERR4161583&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">[</span>
<span class="w">                </span><span class="s2">&quot;/pinky/DATASTORE/Amaranthus_palmeri_PRJEB38372/ERR4161583_1.fastq.gz&quot;</span><span class="p">,</span>
<span class="w">                </span><span class="s2">&quot;/pinky/DATASTORE/Amaranthus_palmeri_PRJEB38372/ERR4161583_2.fastq.gz&quot;</span>
<span class="w">            </span><span class="p">],</span>
<span class="w">            </span><span class="nt">&quot;ERR4161584&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">[</span>
<span class="w">                </span><span class="s2">&quot;/pinky/DATASTORE/Amaranthus_palmeri_PRJEB38372/ERR4161584_1.fastq.gz&quot;</span><span class="p">,</span>
<span class="w">                </span><span class="s2">&quot;/pinky/DATASTORE/Amaranthus_palmeri_PRJEB38372/ERR4161584_2.fastq.gz&quot;</span>
<span class="w">            </span><span class="p">]</span>
<span class="w">        </span><span class="p">},</span>
<span class="w">        </span><span class="nt">&quot;commands&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{}</span>
<span class="w">    </span><span class="p">},</span>
<span class="p">}</span>
</code></pre></div>
</details>
<h3 id="read-trimming-optional">Read trimming (optional)<a class="headerlink" href="#read-trimming-optional" title="Permanent link">&para;</a></h3>
<p>You <em>can</em> perform read trimming on your own before using kmerkit, but we also
provide the option to trim, filter, or subsample your reads within kmerkit
by calling the program <code>fastp</code>. Trimmed read files are written to the workdir, 
and subsequent modules (e.g., <code>count</code> and <code>extract</code>) will 
use the trimmed reads instead of the raw reads (both sets of filepaths can 
be viewed in the project JSON file).</p>
<div class="highlight"><pre><span></span><code>kmerkit<span class="w"> </span>trim<span class="w"> </span>-j<span class="w"> </span>/tmp/dioecy.json<span class="w"> </span>--workers<span class="w"> </span><span class="m">4</span>
</code></pre></div>
<p>Like before we can access the results from the JSON file by calling the
<code>stats</code> module. In this case I also provide a module name (trim) to 
select only results for this module. </p>
<details class="tip">
<summary>kmerkit stats output</summary>
<div class="highlight"><pre><span></span><code>kmerkit<span class="w"> </span>stats<span class="w"> </span>-j<span class="w"> </span>/tmp/dioecy.json<span class="w"> </span>trim
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="err">...</span>
</code></pre></div>
</details>
<h3 id="count-kmers">Count kmers<a class="headerlink" href="#count-kmers" title="Permanent link">&para;</a></h3>
<p>Kmers are counted for each sample using 
<a href="https://github.com/refresh-bio/KMC">kmc</a> at the specified kmer-size. 
Kmers occurring above or below the specified thresholds will be excluded. 
See the <code>kmerkit count</code> page for details on parallelization and memory
consumption for optimizing the speed of this step, which is usually 
the most time consuming. </p>
<div class="highlight"><pre><span></span><code>kmerkit<span class="w"> </span>count<span class="w"> </span>-j<span class="w"> </span>/tmp/dioecy.json<span class="w"> </span>--kmer-size<span class="w"> </span><span class="m">35</span><span class="w"> </span>--min-depth<span class="w"> </span><span class="m">5</span><span class="w"> </span>--workers<span class="w"> </span><span class="m">4</span><span class="w"> </span>
</code></pre></div>
<details class="tip">
<summary>kmerkit stats output</summary>
<div class="highlight"><pre><span></span><code>kmerkit<span class="w"> </span>stats<span class="w"> </span>-j<span class="w"> </span>/tmp/dioecy.json<span class="w"> </span>count
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="go">...</span>
</code></pre></div>
</details>
<h3 id="filter-kmers">Filter kmers<a class="headerlink" href="#filter-kmers" title="Permanent link">&para;</a></h3>
<p>Apply filters to identify target kmers that are enriched in one group of 
samples versus another. In this case, we aim to identify male-specific kmers,
meaning those that are present in males but not females. This can be done 
by setting a high <code>min-map</code> for group 1 (kmers must be present in group 1) and 
setting a low <code>max-map</code> for group 0 (kmers cannot be present in group 0). We
must also assign samples to groups 0 or 1. For studies with many samples this
is most easily done by entering a CSV file (see the kfilter docs section). Here
because there are few samples I use the simpler option of entering the sample
names directly using the <code>-0</code> and <code>-1</code> options to assign to their respective
groups. The min_map and max_map entries each take two ordered values, assigned
to group 0 and 1, respectively.</p>
<div class="highlight"><pre><span></span><code>kmerkit<span class="w"> </span>filter<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--json<span class="w"> </span>/tmp/dioecy.json<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-1<span class="w"> </span><span class="s1">&#39;ERR4161581&#39;</span><span class="w"> </span>-1<span class="w"> </span><span class="s1">&#39;ERR4161583&#39;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-0<span class="w"> </span><span class="s1">&#39;ERR4161582&#39;</span><span class="w"> </span>-0<span class="w"> </span><span class="s1">&#39;ERR4161584&#39;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--min_map<span class="w"> </span><span class="m">0</span>.0<span class="w"> </span><span class="m">0</span>.5<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--max_map<span class="w"> </span><span class="m">0</span>.0<span class="w"> </span><span class="m">1</span>.0<span class="w">  </span>
</code></pre></div>
<details class="tip">
<summary>kmerkit filter results</summary>
<div class="highlight"><pre><span></span><code>kmerkit<span class="w"> </span>stats<span class="w"> </span>-j<span class="w"> </span>/tmp/dioecy.json<span class="w"> </span>filter
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="go">...</span>
</code></pre></div>
</details>
<h3 id="extract-reads">Extract reads<a class="headerlink" href="#extract-reads" title="Permanent link">&para;</a></h3>
<p>Now that we've identified a set of target kmers we will extract reads from 
fastq data files that contain these kmers. This is expected to pull out reads
mapping to male-specific regions of the <i>A. palmeri</i> genome. Here you 
can enter new fastq files to extract data from, or enter the names of samples
already in the project database, which will use the (trimmed) fastq data files
referenced in the JSON file. Here I don't enter any sample names, which 
defaults to performing extractions on all 4 samples in the database (we expect
to not recover any reads for the two female pop samples).</p>
<div class="highlight"><pre><span></span><code>kmerkit<span class="w"> </span>extract<span class="w"> </span>--json<span class="w"> </span>/tmp/dioecy.json<span class="w"> </span>--min-kmers-per-read<span class="w"> </span><span class="m">5</span>
</code></pre></div>
<details class="tip">
<summary>kmerkit stats output</summary>
<div class="highlight"><pre><span></span><code><span class="go">...</span>
</code></pre></div>
</details>
<h3 id="assemble-contigs">Assemble contigs<a class="headerlink" href="#assemble-contigs" title="Permanent link">&para;</a></h3>
<p>From the extracted reads created in the last step, we can now assemble contigs
for each sample (or for all samples pooled together) using the <code>assemble</code> 
module. Here we use the default assembler, spades. This creates a number of
output files in the workdir, which can be summarized with <code>stats</code>. </p>
<div class="highlight"><pre><span></span><code>kmerkit<span class="w"> </span>assemble<span class="w"> </span>--json<span class="w"> </span>/tmp/dioecy.json<span class="w"> </span>...
</code></pre></div>
<p>The main result of Neves et al. was the identification of an approximately
2Mb assembled contig representing a large contiguous male-specific genomic
region. The summary here shows the ...</p>
<details class="tip">
<summary>kmerkit stats output</summary>
<div class="highlight"><pre><span></span><code><span class="go">...</span>
</code></pre></div>
</details>
<h2 id="reproducibility">Reproducibility<a class="headerlink" href="#reproducibility" title="Permanent link">&para;</a></h2>
<p>In addition to your scripts that can be used to reproduce your analysis, the 
JSON project file contains a full record of the samples, parameters, and the
order of analysis steps that make up your analysis.</p>
<details class="success">
<summary>kmerkit project JSON file</summary>
<div class="highlight"><pre><span></span><code><span class="go">...</span>
</code></pre></div>
</details>
<h2 id="todo-post-pipeline-analysis-api">TODO: Post-pipeline analysis API<a class="headerlink" href="#todo-post-pipeline-analysis-api" title="Permanent link">&para;</a></h2>
<p>The <code>kmerkit</code> Python API can be used to perform post-pipeline analyses 
in a jupyter notebook. Here we create a plot of </p>
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">kmerkit</span>

<span class="n">project</span> <span class="o">=</span> <span class="n">kmerkit</span><span class="o">.</span><span class="n">load_json</span><span class="p">(</span><span class="s2">&quot;/tmp/dioecy.json&quot;</span><span class="p">)</span>
<span class="o">...</span>
</code></pre></div>


  




                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.instant", "navigation.tracking", "search.highlight", "search.suggest", "content.code.annotate"], "search": "../../assets/javascripts/workers/search.db81ec45.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.6df46069.min.js"></script>
      
        <script src="../../javascripts/extra.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>