
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../../chapter-3/3.2-markov-models/">
      
      
        <link rel="next" href="../6.1-fitch-multiple-traits/">
      
      <link rel="icon" href="/images/favicon.ico">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-9.0.11">
    
    
      
        <title>6.0 - Fitch Parsimony - Phylogenetic Data Science (IN DEVELOPMENT)</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.0d440cfe.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.2505c338.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  


  <script id="__analytics">function __md_analytics(){function n(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],n("js",new Date),n("config","G-HV34HZZD5N"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){this.value&&n("event","search",{search_term:this.value})}),document$.subscribe(function(){var a=document.forms.feedback;if(void 0!==a)for(var e of a.querySelectorAll("[type=submit]"))e.addEventListener("click",function(e){e.preventDefault();var t=document.location.pathname,e=this.getAttribute("data-md-value");n("event","feedback",{page:t,data:e}),a.firstElementChild.disabled=!0;e=a.querySelector(".md-feedback__note [data-md-value='"+e+"']");e&&(e.hidden=!1)}),a.hidden=!1}),location$.subscribe(function(e){n("config","G-HV34HZZD5N",{page_path:e.pathname})})});var e=document.createElement("script");e.async=!0,e.src="https://www.googletagmanager.com/gtag/js?id=G-HV34HZZD5N",document.getElementById("__analytics").insertAdjacentElement("afterEnd",e)}</script>

  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#tree-inference-parsimony-to-be-updated" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Phylogenetic Data Science (IN DEVELOPMENT)" class="md-header__button md-logo" aria-label="Phylogenetic Data Science (IN DEVELOPMENT)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Phylogenetic Data Science (IN DEVELOPMENT)
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              6.0 - Fitch Parsimony
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent=""  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 0-7 7c0 2.38 1.19 4.47 3 5.74V17a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1v-2.26c1.81-1.27 3-3.36 3-5.74a7 7 0 0 0-7-7M9 21a1 1 0 0 0 1 1h4a1 1 0 0 0 1-1v-1H9v1Z"/></svg>
            </label>
          
        
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 1 7 7c0 2.38-1.19 4.47-3 5.74V17a1 1 0 0 1-1 1H9a1 1 0 0 1-1-1v-2.26C6.19 13.47 5 11.38 5 9a7 7 0 0 1 7-7M9 21v-1h6v1a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1m3-17a5 5 0 0 0-5 5c0 2.05 1.23 3.81 3 4.58V16h4v-2.42c1.77-.77 3-2.53 3-4.58a5 5 0 0 0-5-5Z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/eaton-lab/phylogenetic-data-science" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Phylogenetic Data Science (IN DEVELOPMENT)" class="md-nav__button md-logo" aria-label="Phylogenetic Data Science (IN DEVELOPMENT)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Phylogenetic Data Science (IN DEVELOPMENT)
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/eaton-lab/phylogenetic-data-science" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        About/Ethos
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../installation/" class="md-nav__link">
        Installation/Software
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../course/" class="md-nav__link">
        Columbia course
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
      
      
      
        <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
          0 - Coding Bootcamp
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          0 - Coding Bootcamp
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/getting-started/" class="md-nav__link">
        Getting started
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.0-installation-conda/" class="md-nav__link">
        0.0 - Installation (conda)
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.1-jupyter-and-binder/" class="md-nav__link">
        0.1 - Python Jupyter/Binder
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.2-python-types/" class="md-nav__link">
        0.2 - Python types
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.3-python-funcs/" class="md-nav__link">
        0.3 - Python functions
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.4-python-classes/" class="md-nav__link">
        0.4 - Python classes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.5-python-style/" class="md-nav__link">
        0.5 - Python style
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
      
      
      
        <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
          1 - Trees as data
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          1 - Trees as data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.1-introduction/" class="md-nav__link">
        x.1 - Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.2-tree-thinking/" class="md-nav__link">
        x.2 - Tree Thinking
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.3-visualization/" class="md-nav__link">
        x.3 - Visualization
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.4-features-of-trees/" class="md-nav__link">
        1.4 - Tree data structure
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.5-code-node-class/" class="md-nav__link">
        1.5 - Code - Node class
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.6-code-tree-traversal/" class="md-nav__link">
        1.6 - Code - Tree traversal
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.7-code-tree-class/" class="md-nav__link">
        1.7 - Code - Tree class
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.6-exercises.md" class="md-nav__link">
        x.7 - Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
      
      
      
        <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
          2 - Evolutionary models
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          2 - Evolutionary models
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.0-modeling-evolution/" class="md-nav__link">
        2.0 - Modeling evolution
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.1-idealized-population/" class="md-nav__link">
        2.1 - Idealized population
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.2-wright-fisher/" class="md-nav__link">
        2.2 - Wright-Fisher Process
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.3-coalescent/" class="md-nav__link">
        2.3 - The Coalescent
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.4-coalescent-simulation/" class="md-nav__link">
        2.4 - Coalescent simulation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.x-exercises/" class="md-nav__link">
        2.x - Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
      
      
      
        <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
          3 - Mutation Models
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_7">
          <span class="md-nav__icon md-icon"></span>
          3 - Mutation Models
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-3/3.x-DNA-Markov.md" class="md-nav__link">
        3.0 - Modeling Evolution
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        3.x - Probabilistic and Generative Models
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-3/3.2-markov-models/" class="md-nav__link">
        3.2 - Markov models
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../3.x-other-models.md" class="md-nav__link">
        3.x - Other models
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../1-reading-list.md" class="md-nav__link">
        Reading and Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
      
      
      
        <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
          4 - Data
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_8">
          <span class="md-nav__icon md-icon"></span>
          4 - Data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../4.1-data-formats.md" class="md-nav__link">
        x.1 - data formats
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../4.2-strings-and-matrices.md" class="md-nav__link">
        x.2 - strings, tables, and matrices
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.x-simulating-traits.md" class="md-nav__link">
        x.x - Code - simulating data/traits
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.x-sequence-alignment.md" class="md-nav__link">
        x.x - Code - sequence alignment
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.x-converting-file-formats.md" class="md-nav__link">
        x.x - Code - converting file formats
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../read, write, parse, find, and clean phylip, fasta, fastq, and vcf data." class="md-nav__link">
        Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
      
      
      
        <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
          5 - Phylogenetic Inference
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_9">
          <span class="md-nav__icon md-icon"></span>
          5 - Phylogenetic Inference
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.0 - introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.1 - history
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.2 - choosing a method
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.3 - ..." class="md-nav__link">
        None
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.4 - ..." class="md-nav__link">
        None
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_10" checked>
      
      
      
        <label class="md-nav__link" for="__nav_10" id="__nav_10_label" tabindex="0">
          6 - Inference (Parsimony)
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_10_label" aria-expanded="true">
        <label class="md-nav__title" for="__nav_10">
          <span class="md-nav__icon md-icon"></span>
          6 - Inference (Parsimony)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          6.0 - Fitch Parsimony
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        6.0 - Fitch Parsimony
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    Learning objectives
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#the-principle-of-parsimony" class="md-nav__link">
    The principle of parsimony
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#maximum-parsimony" class="md-nav__link">
    Maximum parsimony
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parsimony-for-teaching-evolution" class="md-nav__link">
    Parsimony for teaching evolution
  </a>
  
    <nav class="md-nav" aria-label="Parsimony for teaching evolution">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#phylogenetic-data" class="md-nav__link">
    Phylogenetic data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#characters-on-a-tree" class="md-nav__link">
    Characters on a tree
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#reconstruction-by-eye" class="md-nav__link">
    Reconstruction by eye
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#fitch-algorithm" class="md-nav__link">
    Fitch algorithm
  </a>
  
    <nav class="md-nav" aria-label="Fitch algorithm">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#effect-of-rooting" class="md-nav__link">
    Effect of Rooting
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#alternate-trees" class="md-nav__link">
    Alternate Trees
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summary" class="md-nav__link">
    Summary
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../6.1-fitch-multiple-traits/" class="md-nav__link">
        6.1 - Character matrices
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        6.2 - Sankoff Parsimony
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_11" >
      
      
      
        <label class="md-nav__link" for="__nav_11" id="__nav_11_label" tabindex="0">
          7 - Inference (Distance-based)
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_11_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_11">
          <span class="md-nav__icon md-icon"></span>
          7 - Inference (Distance-based)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.0-introduction/" class="md-nav__link">
        7.0 - Distance based trees
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.1-least-squares/" class="md-nav__link">
        7.1 - Code - Least-squares
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.2-upgma/" class="md-nav__link">
        7.2 - Code - UPGMA
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.3-neighbor-joining/" class="md-nav__link">
        7.3 - Code - Neighbor Joining
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.4-applied/" class="md-nav__link">
        7.4 - Applied examples
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.5-new-applications/" class="md-nav__link">
        7.5 - New Applications
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_12" >
      
      
      
        <label class="md-nav__link" for="__nav_12" id="__nav_12_label" tabindex="0">
          8 - Inference (Likelihood)
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_12_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_12">
          <span class="md-nav__icon md-icon"></span>
          8 - Inference (Likelihood)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.0 - Maximum Likelihood
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.1 - Genealogies
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.2 - Sequences
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.3 - Inferring
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Reading and Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_13" >
      
      
      
        <label class="md-nav__link" for="__nav_13" id="__nav_13_label" tabindex="0">
          9 - Multispecies Coalescent
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_13_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_13">
          <span class="md-nav__icon md-icon"></span>
          9 - Multispecies Coalescent
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.0-intro.md" class="md-nav__link">
        9.0 - Species trees
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.1-likelihood.md" class="md-nav__link">
        9.1 - Coalescent likelihood
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.2-genealogical.md" class="md-nav__link">
        9.2 - Genealogical thinking
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.2-genetics.md" class="md-nav__link">
        9.3 - Genetic thinking
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.3-MSC.md" class="md-nav__link">
        9.4 - MSC likelihood
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Reading and Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    Learning objectives
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#the-principle-of-parsimony" class="md-nav__link">
    The principle of parsimony
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#maximum-parsimony" class="md-nav__link">
    Maximum parsimony
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parsimony-for-teaching-evolution" class="md-nav__link">
    Parsimony for teaching evolution
  </a>
  
    <nav class="md-nav" aria-label="Parsimony for teaching evolution">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#phylogenetic-data" class="md-nav__link">
    Phylogenetic data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#characters-on-a-tree" class="md-nav__link">
    Characters on a tree
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#reconstruction-by-eye" class="md-nav__link">
    Reconstruction by eye
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#fitch-algorithm" class="md-nav__link">
    Fitch algorithm
  </a>
  
    <nav class="md-nav" aria-label="Fitch algorithm">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#effect-of-rooting" class="md-nav__link">
    Effect of Rooting
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#alternate-trees" class="md-nav__link">
    Alternate Trees
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summary" class="md-nav__link">
    Summary
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  



<h1 id="tree-inference-parsimony-to-be-updated">Tree Inference (Parsimony) (TO BE UPDATED)<a class="headerlink" href="#tree-inference-parsimony-to-be-updated" title="Permanent link">&para;</a></h1>
<p><span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.62L12 2 9.19 8.62 2 9.24l5.45 4.73L5.82 21 12 17.27Z"/></svg></span> <em>Follow along in a binder jupyter notebook:</em>
<a href="https://mybinder.org/v2/gh/eaton-lab/phylogenetic-data-science/HEAD?filepath=docs%2Fchapter-6%2Fnotebooks%2Fparsimony.ipynb"><span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 7V5h2V4a2 2 0 0 1 2-2h6v7l2.5-1.5L18 9V2h1c1.05 0 2 .95 2 2v16c0 1.05-.95 2-2 2H7c-1.05 0-2-.95-2-2v-1H3v-2h2v-4H3v-2h2V7H3m4 4H5v2h2v-2m0-4V5H5v2h2m0 12v-2H5v2h2Z"/></svg></span> fitch-parsimony.ipynb</a></p>
<h2 id="learning-objectives">Learning objectives<a class="headerlink" href="#learning-objectives" title="Permanent link">&para;</a></h2>
<!-- Before starting this chapter I recommend reading [chapter 5](../6.0-tree-inference.md) for an introduction to phylogenetic inference.  -->
<p>By the end of this chapter you will:</p>
<ol>
<li>Understand how the principle of parsimony is used in phylogenetic tree inference.</li>
<li>Be able to write a function to calculate a parsimony score.</li>
<li>Know the difference between Fitch and Sankoff parsimony algorithms.</li>
</ol>
<h2 id="the-principle-of-parsimony">The principle of parsimony<a class="headerlink" href="#the-principle-of-parsimony" title="Permanent link">&para;</a></h2>
<p>The principle of parsimony, also termed <a href="https://en.wikipedia.org/wiki/Occam's_razor"><em>Occam's razor</em></a>, is widely used in
biology. It advocates that when presented with competing hypotheses about the 
same prediction -- such as different phylogenetic hypotheses for a set of 
samples -- one should select the solution that requires the fewest 
assumptions. In evolutionary terms, this is often interpreted to mean that 
the most <em>parsimonious</em> solution is the one that requires the fewest 
observed evolutionary changes/events.</p>
<h2 id="maximum-parsimony">Maximum parsimony<a class="headerlink" href="#maximum-parsimony" title="Permanent link">&para;</a></h2>
<p>Finding the most parsimonious tree given a phylogenetic dataset thus 
requires, most importantly, an algorithm for counting the number of 
evolutionary changes on a phylogenetic tree. This turns out to be 
relatively simple to code algorithmically, but can be quite complex in
terms of the underlying philosophical and statistical interpretations
of what these changes mean. A subject we will return to later. In this
chapter we will learn algorithmic methods for <em>counting evolutionary changes</em>, 
by introducing two algorithms, <em>Fitch parsimony</em> and <em>Sankoff parsimony</em>.</p>
<p>The number of evolutionary changes that must occur on a tree, given a 
dataset of character states at the tips, can be referred to as its 
<em>parsimony score</em>. Given a method for computing parsimony scores, one could
in theory apply it to every possible tree topology to N taxa to find the
one or more trees that require the minimum number of changes. Because the
total number of possible trees is typically far too large to compute on, 
<a href=".">heuristic search</a> methods are employed instead, which score only a subset 
of trees. This general approach, of comparing parsimony scores 
on many trees to find the best scoring tree is termed <em>maximum parsimony (MP)</em>.</p>
<h2 id="parsimony-for-teaching-evolution">Parsimony for teaching evolution<a class="headerlink" href="#parsimony-for-teaching-evolution" title="Permanent link">&para;</a></h2>
<p>Because parsimony is simple, it is frequently used in introductory evolution
textbooks, and other resources, to show evidence for evolution without the
need for describing more complex statistical models. For example, we could
compare two hypotheses for the evolution of relationships among vertebrate
lineages using a number of discrete character traits to show that one tree
requires more evolutionary events than the other. In hypothesis 1 the 
evolution of a <em>bony skeleton</em> is inferred to have occurred only once, 
whereas in hypothesis 2 it is inferred to have evolved twice independently.</p>
<!-- A convergent evolutionary pattern that called *homoplasy*.  -->

<ul>
<li><mark>add textbook style images here of two trees with apomorphies, one with homoplasy.</mark></li>
</ul>
<h3 id="phylogenetic-data">Phylogenetic data<a class="headerlink" href="#phylogenetic-data" title="Permanent link">&para;</a></h3>
<p>As with most phylogenetic inference algorithms, parsimony is intended to 
apply to discrete character measurements, such as binary data 
(<code>0</code> or <code>1</code>) or multistate characters (e.g., <code>A</code>, <code>B</code>, <code>C</code> or <code>D</code>). We discuss
discrete data in much detail in <a href=".">chapter 3</a>. Parsimony can be applied to DNA
sequence data, and is frequently used alongside other methods for analyzing
DNA, such as likelihood and distance-based inference methods. However,
the <em>niche</em> where maximum parsimony remains most commonly used today, is in 
the analysis of morphological datasets, especially in paleontology,
where DNA sequence data is often not available.</p>
<ul>
<li><mark>describe assumptions about traits used in the image above, such as 
the number of possible states.</mark></li>
<li><mark>describe that there is active debate about the continued value of 
morphological data for <em>inferring trees</em>, as opposed to just being studied
on trees inferred with other data. Link to readings.</mark></li>
</ul>
<h3 id="characters-on-a-tree">Characters on a tree<a class="headerlink" href="#characters-on-a-tree" title="Permanent link">&para;</a></h3>
<p>Let's start with a simple example where we have measured a single
discrete binary character for five different species, and we want to 
examine this character with respect to a particular phylogenetic 
hypothesis. We will make the assumption that the character can only be in
one of two possible states, and thus that the only possible evolutionary
changes are from <code>0 -&gt; 1</code> or <code>1 -&gt; 0</code>. We only have data for the tip
samples, but from these data we can infer the minimum number of 
evolutionary changes that must occur, given a tree, for the character states
to be distributed among tips in the way that they are. </p>
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">toytree</span>

<span class="c1"># an example five tip tree (with tip names alpha-epsilon)</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">toytree</span><span class="o">.</span><span class="n">tree</span><span class="p">(</span><span class="s2">&quot;(((alpha,delta),gamma),(beta,epsilon));&quot;</span><span class="p">)</span>

<span class="c1"># a single binary trait character state for each species</span>
<span class="n">trait</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;alpha&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
    <span class="s2">&quot;beta&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s2">&quot;delta&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
    <span class="s2">&quot;epsilon&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s2">&quot;gamma&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
<span class="p">}</span>
</code></pre></div>
<p>We can easily visualize this trait in <code>toytree</code> by setting it as an attribute
of the Nodes, for example by using the <code>set_node_data</code> ToyTree function. In 
this example we store the trait as a feature named <code>trait</code>. We can then get 
and show the data for this feature in a dataframe using <code>get_node_data</code>. 
This dataframe shows node <code>idx</code> labels in the row index, and the values for
each selected Node feature (here <code>name</code> and <code>trait</code>). As we can see in this
table, and in the plots below, there are four internal Nodes that do not
have a name or trait values assigned.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># sets a .trait attribute to each Node in trait dictionary</span>
<span class="n">tree</span><span class="o">.</span><span class="n">set_node_data</span><span class="p">(</span>
    <span class="n">feature</span><span class="o">=</span><span class="s2">&quot;trait&quot;</span><span class="p">,</span>
    <span class="n">mapping</span><span class="o">=</span><span class="n">trait</span><span class="p">,</span>
    <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># get and display the Node data as a dataframe</span>
<span class="n">tree</span><span class="o">.</span><span class="n">get_node_data</span><span class="p">([</span><span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;trait&#39;</span><span class="p">])</span>
</code></pre></div>
<figure>
<table>
<thead>
<tr>
<th align="left"></th>
<th align="left">name</th>
<th align="left">trait</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">0</td>
<td align="left">alpha</td>
<td align="left">1</td>
</tr>
<tr>
<td align="left">1</td>
<td align="left">delta</td>
<td align="left">1</td>
</tr>
<tr>
<td align="left">2</td>
<td align="left">gamma</td>
<td align="left">1</td>
</tr>
<tr>
<td align="left">3</td>
<td align="left">beta</td>
<td align="left">0</td>
</tr>
<tr>
<td align="left">4</td>
<td align="left">epsilon</td>
<td align="left">0</td>
</tr>
<tr>
<td align="left">5</td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="left">6</td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="left">7</td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="left">8</td>
<td align="left"></td>
<td align="left"></td>
</tr>
</tbody>
</table>
</figure>
<p>Our current knowledge is only of the data at the tips, which looks like
this on our tree hypothesis:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># draw tree with tip Nodes colored by feature &#39;trait&#39; </span>
<span class="n">tree</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span>
    <span class="n">tree_style</span><span class="o">=</span><span class="s1">&#39;s&#39;</span><span class="p">,</span>
    <span class="n">use_edge_lengths</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> 
    <span class="n">tip_labels_style</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;font-size&quot;</span><span class="p">:</span> <span class="mi">18</span><span class="p">},</span>
    <span class="n">node_labels</span><span class="o">=</span><span class="s2">&quot;trait&quot;</span><span class="p">,</span>
    <span class="n">node_colors</span><span class="o">=</span><span class="p">[</span>
        <span class="n">toytree</span><span class="o">.</span><span class="n">color</span><span class="o">.</span><span class="n">COLORS1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span><span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="s2">&quot;lightgrey&quot;</span> 
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">get_node_data</span><span class="p">(</span><span class="s2">&quot;trait&quot;</span><span class="p">)</span>
    <span class="p">],</span>
<span class="p">);</span>
</code></pre></div>
<figure>
<p><img _="," alt="Heatmap image" loading="lazy" src="../../chapter-1/notebooks/fig-6.1.svg" width="305" />
  </p>
<figcaption>Fig. 6.1</figcaption>
</figure>
<h3 id="reconstruction-by-eye">Reconstruction by eye<a class="headerlink" href="#reconstruction-by-eye" title="Permanent link">&para;</a></h3>
<p>Given that this is a fairly small tree, we can try to estimate 
the number of evolutionary changes by eye, by counting the number
of events that would be required to explain the data at the tips. This is 
similar to what was done in the vertebrate tree example above.
Try doing this now. </p>
<p>You should find that the minimum number of changes to explain this data is 1.
Depending on where that change occurred in the past, there are two equally
good reconstructions for this character on this tree topology. Either the 
root state was <code>1</code> and transitioned to <code>0</code> in the common ancestor of 
(<em>beta, epsilon</em>); or, the root state <code>0</code> and transitioned to <code>1</code> in the 
common ancestor of (<em>alpha, delta, gamma</em>).</p>
<figure>
<p><img _="," alt="Heatmap image" loading="lazy" src="../../chapter-1/notebooks/fig-6.2.svg" width="600" />
  </p>
<figcaption>Fig. 6.2</figcaption>
</figure>
<h2 id="fitch-algorithm">Fitch algorithm<a class="headerlink" href="#fitch-algorithm" title="Permanent link">&para;</a></h2>
<ul>
<li>
<p>description of Fitch...</p>
</li>
<li>
<p>set intersection</p>
</li>
<li>
<p>set union</p>
</li>
</ul>
<p>Like we have done before, we will begin with a very simple implementation
of the Fitch algorithm for computing parsimony scores, and build to a 
more complex version. This is intended both to demonstrate how the algorithm
works as clearly as possible, as well as to demonstrate some common coding
practices that will be helpful for coding similar algorithms. </p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Any</span>

<span class="k">def</span> <span class="nf">fitch_parsimony_single</span><span class="p">(</span><span class="n">tree</span><span class="p">:</span> <span class="n">toytree</span><span class="o">.</span><span class="n">ToyTree</span><span class="p">,</span> <span class="n">trait</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span><span class="n">Any</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return Fitch parsimony score given a tree and single trait.</span>

<span class="sd">    For didactic purposes this function will also store a feature named </span>
<span class="sd">    &#39;fitch&#39; to every Node which can be examined/visualized afterwards.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    tree: ToyTree</span>
<span class="sd">        A tree on which to count state changes.</span>
<span class="sd">    trait: Dict</span>
<span class="sd">        A dict mapping tip names to a single discrete character state.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        The minimum changes required for trait data to evolve on this tree.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># counter to keep track of change events</span>
    <span class="n">nchanges</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1"># iterate over Nodes in idxorder (postorder sorted) traversal</span>
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree</span><span class="p">:</span>

        <span class="c1"># leaves are visited first, and converted to a set type</span>
        <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">():</span>
            <span class="n">node</span><span class="o">.</span><span class="n">fitch</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">trait</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">])</span>

        <span class="c1"># internal Nodes examine the sets of their children&#39;s states</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">node</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">():</span>

            <span class="c1"># check for shared (intersecting) states</span>
            <span class="n">shared</span> <span class="o">=</span> <span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">fitch</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="p">))</span>

            <span class="c1"># if any states are shared then ancestor inherits this state</span>
            <span class="k">if</span> <span class="n">shared</span><span class="p">:</span>
                <span class="n">node</span><span class="o">.</span><span class="n">fitch</span> <span class="o">=</span> <span class="n">shared</span>

            <span class="c1"># if none shared, then store the union and increment counter</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">node</span><span class="o">.</span><span class="n">fitch</span> <span class="o">=</span> <span class="nb">set</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="o">*</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">fitch</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="p">))</span>
                <span class="n">nchanges</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">nchanges</span>
</code></pre></div>
<p>Applying this function to our example trait dataset returns a score of 1, 
just as we expected from examining the tree above. 
<div class="highlight"><pre><span></span><code><span class="n">fitch_parsimony_single</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">trait</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="c1"># 1</span>
</code></pre></div></p>
<p>Because our function was coded to also record the set of states at each
Node during the tree traversal, we can visualize this after running the
function to ensure that it performed as we expected. This data was stored
to a feature named "fitch" on each Node, and so we can extract it easily
with <code>toytree</code> by entering this feature name to the <code>node_labels</code> draw
argument.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># draw tree to show Fitch parsimony sets at each Node as rectangle markers.</span>
<span class="n">tree</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span>
    <span class="n">width</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> 
    <span class="n">use_edge_lengths</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">node_sizes</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> 
    <span class="n">node_markers</span><span class="o">=</span><span class="s1">&#39;r2x1&#39;</span><span class="p">,</span>
    <span class="n">node_mask</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">node_colors</span><span class="o">=</span><span class="s2">&quot;lightgrey&quot;</span><span class="p">,</span>
    <span class="n">node_labels</span><span class="o">=</span><span class="s2">&quot;fitch&quot;</span><span class="p">,</span> 
    <span class="n">node_labels_style</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;font-size&#39;</span><span class="p">:</span> <span class="mi">15</span><span class="p">},</span>
    <span class="n">tip_labels_style</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;font-size&#39;</span><span class="p">:</span> <span class="mi">15</span><span class="p">,</span> <span class="s1">&#39;-toyplot-anchor-shift&#39;</span><span class="p">:</span> <span class="mi">30</span><span class="p">}</span>
<span class="p">);</span>
</code></pre></div>
<figure>
<p><img _="," alt="Heatmap image" loading="lazy" src="../../chapter-1/notebooks/fig-6.3.svg" width="500" />
  </p>
<figcaption>Fig. 6.3</figcaption>
</figure>
<h3 id="effect-of-rooting">Effect of Rooting<a class="headerlink" href="#effect-of-rooting" title="Permanent link">&para;</a></h3>
<p>As mentioned earlier, although this algorithm is applied in a postorder 
traversal (from tips to the root), it does not actually rely on information
about the rooting to count the number of changes. In fact, we could 
arbitrarily root the tree on any edge and it would return the same score. 
Let's verify this now by re-rooting the tree on each possible edge and 
printing the Fitch parsimony score.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># iterate over each edge that can become new root</span>
<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree</span><span class="p">:</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">node</span><span class="o">.</span><span class="n">is_root</span><span class="p">():</span>
        <span class="c1"># get a copy of tree rooted on new edge (selected by Node below)</span>
        <span class="n">new_rooted_tree</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">root</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>

        <span class="c1"># print the parsimony score of this new tree</span>
        <span class="n">score</span> <span class="o">=</span> <span class="n">fitch_parsimony_single</span><span class="p">(</span><span class="n">new_rooted_tree</span><span class="p">,</span> <span class="n">trait</span><span class="p">)</span>
        <span class="n">clade</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">get_leaf_names</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;score=</span><span class="si">{</span><span class="n">score</span><span class="si">}</span><span class="s1"> when rooted on </span><span class="si">{</span><span class="n">clade</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="c1"># score=1 when rooted on [&#39;alpha&#39;]</span>
<span class="c1"># score=1 when rooted on [&#39;delta&#39;]</span>
<span class="c1"># score=1 when rooted on [&#39;gamma&#39;]</span>
<span class="c1"># score=1 when rooted on [&#39;beta&#39;]</span>
<span class="c1"># score=1 when rooted on [&#39;epsilon&#39;]</span>
<span class="c1"># score=1 when rooted on [&#39;alpha&#39;, &#39;delta&#39;]</span>
<span class="c1"># score=1 when rooted on [&#39;alpha&#39;, &#39;delta&#39;, &#39;gamma&#39;]</span>
<span class="c1"># score=1 when rooted on [&#39;beta&#39;, &#39;epsilon&#39;]</span>
</code></pre></div>
<p>Once again we can visualize the sets of one of these re-rooted trees to 
see that regardless of the rooting, this tree still requires a minimum 
of only one event to explain the data at the tips.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># root original tree on edge above (alpha, delta)</span>
<span class="n">new_rooted_tree</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">root</span><span class="p">(</span><span class="s1">&#39;alpha&#39;</span><span class="p">,</span> <span class="s1">&#39;delta&#39;</span><span class="p">)</span>

<span class="c1"># get Fitch set reconstruction on Nodes</span>
<span class="n">fitch_parsimony_single</span><span class="p">(</span><span class="n">new_rooted_tree</span><span class="p">,</span> <span class="n">trait</span><span class="p">)</span>

<span class="c1"># draw the tree to show Node sets</span>
<span class="n">new_rooted_tree</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span>
    <span class="n">width</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> 
    <span class="n">use_edge_lengths</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">node_sizes</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> 
    <span class="n">node_markers</span><span class="o">=</span><span class="s1">&#39;r2x1&#39;</span><span class="p">,</span>
    <span class="n">node_mask</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">node_colors</span><span class="o">=</span><span class="s2">&quot;lightgrey&quot;</span><span class="p">,</span>
    <span class="n">node_labels</span><span class="o">=</span><span class="s2">&quot;fitch&quot;</span><span class="p">,</span> 
    <span class="n">node_labels_style</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;font-size&#39;</span><span class="p">:</span> <span class="mi">15</span><span class="p">},</span>
    <span class="n">tip_labels_style</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;font-size&#39;</span><span class="p">:</span> <span class="mi">15</span><span class="p">,</span> <span class="s1">&#39;-toyplot-anchor-shift&#39;</span><span class="p">:</span> <span class="mi">30</span><span class="p">}</span>
<span class="p">);</span>
</code></pre></div>
<figure>
<p><img _="," alt="Heatmap image" loading="lazy" src="../../chapter-1/notebooks/fig-6.4.svg" width="500" />
  </p>
<figcaption>Fig. 6.4</figcaption>
</figure>
<h3 id="alternate-trees">Alternate Trees<a class="headerlink" href="#alternate-trees" title="Permanent link">&para;</a></h3>
<p>When applied to alternative tree hypotheses the parsimony score may be 
higher or lower depending on their fit to the data. The example below
has a higher Fitch parsimony score (worse fit to the data). </p>
<p><div class="highlight"><pre><span></span><code><span class="c1"># get an alternate tree topology</span>
<span class="n">tree2</span> <span class="o">=</span> <span class="n">toytree</span><span class="o">.</span><span class="n">tree</span><span class="p">(</span><span class="s2">&quot;(((alpha,beta),gamma),(delta,epsilon));&quot;</span><span class="p">)</span>

<span class="c1"># print the fitch score</span>
<span class="n">score</span> <span class="o">=</span> <span class="n">fitch_parsimony_single</span><span class="p">(</span><span class="n">tree2</span><span class="p">,</span> <span class="n">trait</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;score=</span><span class="si">{</span><span class="n">score</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># draw the tree showing sets on the Node labels.</span>
<span class="n">tree2</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span>
    <span class="n">width</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> 
    <span class="n">use_edge_lengths</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">node_sizes</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> 
    <span class="n">node_markers</span><span class="o">=</span><span class="s1">&#39;r2x1&#39;</span><span class="p">,</span>
    <span class="n">node_mask</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">node_colors</span><span class="o">=</span><span class="s2">&quot;lightgrey&quot;</span><span class="p">,</span>
    <span class="n">node_labels</span><span class="o">=</span><span class="s2">&quot;fitch&quot;</span><span class="p">,</span> 
    <span class="n">node_labels_style</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;font-size&#39;</span><span class="p">:</span> <span class="mi">15</span><span class="p">},</span>
    <span class="n">tip_labels_style</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;font-size&#39;</span><span class="p">:</span> <span class="mi">15</span><span class="p">,</span> <span class="s1">&#39;-toyplot-anchor-shift&#39;</span><span class="p">:</span> <span class="mi">30</span><span class="p">}</span>
<span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="c1"># score=2</span>
</code></pre></div></p>
<figure>
<p><img _="," alt="Heatmap image" loading="lazy" src="../../chapter-1/notebooks/fig-6.5.svg" width="500" />
  </p>
<figcaption>Fig. 6.5</figcaption>
</figure>
<h2 id="summary">Summary<a class="headerlink" href="#summary" title="Permanent link">&para;</a></h2>
<p>...</p>


  




                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.instant", "navigation.tracking", "search.highlight", "search.suggest", "content.code.annotate"], "search": "../../assets/javascripts/workers/search.db81ec45.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.6df46069.min.js"></script>
      
        <script src="../../javascripts/extra.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>