
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../6.0-fitch-parsimony/">
      
      
        <link rel="next" href="../../chapter-7/7.0-introduction/">
      
      <link rel="icon" href="/images/favicon.ico">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-9.0.11">
    
    
      
        <title>6.1 - Character matrices - Phylogenetic Data Science (IN DEVELOPMENT)</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.0d440cfe.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.2505c338.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  


  <script id="__analytics">function __md_analytics(){function n(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],n("js",new Date),n("config","G-HV34HZZD5N"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){this.value&&n("event","search",{search_term:this.value})}),document$.subscribe(function(){var a=document.forms.feedback;if(void 0!==a)for(var e of a.querySelectorAll("[type=submit]"))e.addEventListener("click",function(e){e.preventDefault();var t=document.location.pathname,e=this.getAttribute("data-md-value");n("event","feedback",{page:t,data:e}),a.firstElementChild.disabled=!0;e=a.querySelector(".md-feedback__note [data-md-value='"+e+"']");e&&(e.hidden=!1)}),a.hidden=!1}),location$.subscribe(function(e){n("config","G-HV34HZZD5N",{page_path:e.pathname})})});var e=document.createElement("script");e.async=!0,e.src="https://www.googletagmanager.com/gtag/js?id=G-HV34HZZD5N",document.getElementById("__analytics").insertAdjacentElement("afterEnd",e)}</script>

  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-objectives" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Phylogenetic Data Science (IN DEVELOPMENT)" class="md-header__button md-logo" aria-label="Phylogenetic Data Science (IN DEVELOPMENT)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Phylogenetic Data Science (IN DEVELOPMENT)
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              6.1 - Character matrices
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent=""  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 0-7 7c0 2.38 1.19 4.47 3 5.74V17a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1v-2.26c1.81-1.27 3-3.36 3-5.74a7 7 0 0 0-7-7M9 21a1 1 0 0 0 1 1h4a1 1 0 0 0 1-1v-1H9v1Z"/></svg>
            </label>
          
        
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 1 7 7c0 2.38-1.19 4.47-3 5.74V17a1 1 0 0 1-1 1H9a1 1 0 0 1-1-1v-2.26C6.19 13.47 5 11.38 5 9a7 7 0 0 1 7-7M9 21v-1h6v1a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1m3-17a5 5 0 0 0-5 5c0 2.05 1.23 3.81 3 4.58V16h4v-2.42c1.77-.77 3-2.53 3-4.58a5 5 0 0 0-5-5Z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/eaton-lab/phylogenetic-data-science" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Phylogenetic Data Science (IN DEVELOPMENT)" class="md-nav__button md-logo" aria-label="Phylogenetic Data Science (IN DEVELOPMENT)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Phylogenetic Data Science (IN DEVELOPMENT)
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/eaton-lab/phylogenetic-data-science" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        About/Ethos
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../installation/" class="md-nav__link">
        Installation/Software
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../course/" class="md-nav__link">
        Columbia course
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
      
      
      
        <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
          0 - Coding Bootcamp
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          0 - Coding Bootcamp
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/getting-started/" class="md-nav__link">
        Getting started
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.0-installation-conda/" class="md-nav__link">
        0.0 - Installation (conda)
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.1-jupyter-and-binder/" class="md-nav__link">
        0.1 - Python Jupyter/Binder
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.2-python-types/" class="md-nav__link">
        0.2 - Python types
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.3-python-funcs/" class="md-nav__link">
        0.3 - Python functions
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.4-python-classes/" class="md-nav__link">
        0.4 - Python classes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../bootcamp/0.5-python-style/" class="md-nav__link">
        0.5 - Python style
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
      
      
      
        <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
          1 - Trees as data
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          1 - Trees as data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.1-introduction/" class="md-nav__link">
        x.1 - Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.2-tree-thinking/" class="md-nav__link">
        x.2 - Tree Thinking
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.3-visualization/" class="md-nav__link">
        x.3 - Visualization
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.4-features-of-trees/" class="md-nav__link">
        1.4 - Tree data structure
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.5-code-node-class/" class="md-nav__link">
        1.5 - Code - Node class
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.6-code-tree-traversal/" class="md-nav__link">
        1.6 - Code - Tree traversal
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.7-code-tree-class/" class="md-nav__link">
        1.7 - Code - Tree class
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-1/1.6-exercises.md" class="md-nav__link">
        x.7 - Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
      
      
      
        <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
          2 - Evolutionary models
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          2 - Evolutionary models
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.0-modeling-evolution/" class="md-nav__link">
        2.0 - Modeling evolution
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.1-idealized-population/" class="md-nav__link">
        2.1 - Idealized population
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.2-wright-fisher/" class="md-nav__link">
        2.2 - Wright-Fisher Process
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.3-coalescent/" class="md-nav__link">
        2.3 - The Coalescent
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.4-coalescent-simulation/" class="md-nav__link">
        2.4 - Coalescent simulation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-2/2.x-exercises/" class="md-nav__link">
        2.x - Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
      
      
      
        <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
          3 - Mutation Models
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_7">
          <span class="md-nav__icon md-icon"></span>
          3 - Mutation Models
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-3/3.x-DNA-Markov.md" class="md-nav__link">
        3.0 - Modeling Evolution
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        3.x - Probabilistic and Generative Models
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-3/3.2-markov-models/" class="md-nav__link">
        3.2 - Markov models
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../3.x-other-models.md" class="md-nav__link">
        3.x - Other models
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../1-reading-list.md" class="md-nav__link">
        Reading and Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
      
      
      
        <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
          4 - Data
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_8">
          <span class="md-nav__icon md-icon"></span>
          4 - Data
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../4.1-data-formats.md" class="md-nav__link">
        x.1 - data formats
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../4.2-strings-and-matrices.md" class="md-nav__link">
        x.2 - strings, tables, and matrices
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.x-simulating-traits.md" class="md-nav__link">
        x.x - Code - simulating data/traits
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.x-sequence-alignment.md" class="md-nav__link">
        x.x - Code - sequence alignment
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.x-converting-file-formats.md" class="md-nav__link">
        x.x - Code - converting file formats
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../read, write, parse, find, and clean phylip, fasta, fastq, and vcf data." class="md-nav__link">
        Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
      
      
      
        <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
          5 - Phylogenetic Inference
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_9">
          <span class="md-nav__icon md-icon"></span>
          5 - Phylogenetic Inference
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.0 - introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.1 - history
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.2 - choosing a method
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.3 - ..." class="md-nav__link">
        None
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../x.4 - ..." class="md-nav__link">
        None
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_10" checked>
      
      
      
        <label class="md-nav__link" for="__nav_10" id="__nav_10_label" tabindex="0">
          6 - Inference (Parsimony)
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_10_label" aria-expanded="true">
        <label class="md-nav__title" for="__nav_10">
          <span class="md-nav__icon md-icon"></span>
          6 - Inference (Parsimony)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../6.0-fitch-parsimony/" class="md-nav__link">
        6.0 - Fitch Parsimony
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          6.1 - Character matrices
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        6.1 - Character matrices
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    Learning objectives
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#character-matrix" class="md-nav__link">
    Character matrix
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-organization" class="md-nav__link">
    Data Organization
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#array-operations" class="md-nav__link">
    Array Operations
  </a>
  
    <nav class="md-nav" aria-label="Array Operations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pandas-vs-numpy-speed" class="md-nav__link">
    Pandas vs Numpy speed
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#boolean-masking" class="md-nav__link">
    Boolean masking
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#array-indexing" class="md-nav__link">
    Array indexing
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#unique-patterns" class="md-nav__link">
    Unique patterns
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#fitch-parsimony-function" class="md-nav__link">
    Fitch Parsimony Function
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        6.2 - Sankoff Parsimony
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_11" >
      
      
      
        <label class="md-nav__link" for="__nav_11" id="__nav_11_label" tabindex="0">
          7 - Inference (Distance-based)
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_11_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_11">
          <span class="md-nav__icon md-icon"></span>
          7 - Inference (Distance-based)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.0-introduction/" class="md-nav__link">
        7.0 - Distance based trees
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.1-least-squares/" class="md-nav__link">
        7.1 - Code - Least-squares
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.2-upgma/" class="md-nav__link">
        7.2 - Code - UPGMA
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.3-neighbor-joining/" class="md-nav__link">
        7.3 - Code - Neighbor Joining
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.4-applied/" class="md-nav__link">
        7.4 - Applied examples
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-7/7.5-new-applications/" class="md-nav__link">
        7.5 - New Applications
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_12" >
      
      
      
        <label class="md-nav__link" for="__nav_12" id="__nav_12_label" tabindex="0">
          8 - Inference (Likelihood)
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_12_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_12">
          <span class="md-nav__icon md-icon"></span>
          8 - Inference (Likelihood)
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.0 - Maximum Likelihood
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.1 - Genealogies
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.2 - Sequences
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        x.3 - Inferring
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Reading and Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_13" >
      
      
      
        <label class="md-nav__link" for="__nav_13" id="__nav_13_label" tabindex="0">
          9 - Multispecies Coalescent
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_13_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_13">
          <span class="md-nav__icon md-icon"></span>
          9 - Multispecies Coalescent
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.0-intro.md" class="md-nav__link">
        9.0 - Species trees
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.1-likelihood.md" class="md-nav__link">
        9.1 - Coalescent likelihood
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.2-genealogical.md" class="md-nav__link">
        9.2 - Genealogical thinking
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.2-genetics.md" class="md-nav__link">
        9.3 - Genetic thinking
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../chapter-9/9.3-MSC.md" class="md-nav__link">
        9.4 - MSC likelihood
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Reading and Exercises
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    Learning objectives
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#character-matrix" class="md-nav__link">
    Character matrix
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-organization" class="md-nav__link">
    Data Organization
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#array-operations" class="md-nav__link">
    Array Operations
  </a>
  
    <nav class="md-nav" aria-label="Array Operations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pandas-vs-numpy-speed" class="md-nav__link">
    Pandas vs Numpy speed
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#boolean-masking" class="md-nav__link">
    Boolean masking
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#array-indexing" class="md-nav__link">
    Array indexing
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#unique-patterns" class="md-nav__link">
    Unique patterns
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#fitch-parsimony-function" class="md-nav__link">
    Fitch Parsimony Function
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  



  <h1>6.1 - Character matrices</h1>

<h2 id="learning-objectives">Learning objectives<a class="headerlink" href="#learning-objectives" title="Permanent link">&para;</a></h2>
<p>In this lesson you will:</p>
<ul>
<li>Learn methods for filtering a two dimensional character array with <code>numpy</code>.</li>
<li>Write a function to calculate Fitch parsimony score for multiple traits.</li>
<li></li>
</ul>
<h2 id="character-matrix">Character matrix<a class="headerlink" href="#character-matrix" title="Permanent link">&para;</a></h2>
<p>Most phylogenetic datasets will contain measurements of many traits, and 
so let's now expand our Fitch parsimony algorithm to accomodate such data, 
where the returned score will represent the sum of events across each measured 
trait. A collection of multiple trait measurements is typically stored in a
2-dimensional <em>matrix</em> (also termed an <em>array</em>). An example character matrix 
is shown below, and in Python would typically be stored either as 
a <code>numpy.ndarray</code>, or as a <code>pandas.DataFrame</code>. </p>
<p>This example includes seven binary characters measured across five taxa. 
Another common type of data matrix is a sequence alignment, which can often 
span many thousands of columns, where each can take one of four possible 
states (A, C, G, T), or sometimes even ambiguous states (recorded as IUPAC 
character R, S, Y, W, M). Typically, the rows of a character matrix will
represent the taxa, and the columns are each a different trait; however,
this is not always the case, and so it is useful to examine your data and
check for row and column labels.</p>
<!-- A data matrix 
could also represent many traits with different numbers of character states, 
such as with morphological measurements, in which case it can be important to 
also record the number of allowable states for each trait. -->
<p>In this lesson, because we are still focused on the topic of parsimony tree 
inference, we will focus only on discrete characters, and not on quantitative 
character matrices. Ideally, the Fitch algorithm that we design below should 
work properly to return a score for any discrete character matrix. Let's 
start with an example using a tree and the binary character matrix below.</p>
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">toytree</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="c1"># example five tip tree (with tip names alpha-epsilon)</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">toytree</span><span class="o">.</span><span class="n">tree</span><span class="p">(</span><span class="s2">&quot;(((alpha,delta),gamma),(beta,epsilon));&quot;</span><span class="p">)</span>

<span class="c1"># example character array</span>
<span class="n">arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span>
    <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span> <span class="p">,</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
    <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="p">])</span>
</code></pre></div>
<h2 id="data-organization">Data Organization<a class="headerlink" href="#data-organization" title="Permanent link">&para;</a></h2>
<p>In the last notebook we used the <code>ToyTree.set_node_data</code> function to set 
character data as attributes of Nodes of a tree, which provided an easy 
way to keep data correctly associated with the proper taxon names. In the 
example above, you can see that the data matrix stored as a <code>numpy.array</code> 
is less clear, and could be interpreted ambiguously about which rows 
correspond to which taxon names. To reduce errors, it is good practice to 
retain clarity about the assignment of data to tips in a tree, since the 
order of tips is changeable, and often arbitrary. </p>
<p>One way to more clearly label an array is to wrap it as a <code>pandas.DataFrame</code>
object, like below. This has similar functionality to a <code>numpy.ndarray</code>, and
in fact stores the data as an array under the hood, but it also includes
labels on the rows and columns. These labels can be used to index items
from the dataframe by row (index) and column names, as well as by indexing
with numeric indices, like in <code>numpy</code>. <code>pandas</code> has several convenient 
functions for reading and writing data from CSV and other tabular formatted 
data files, and thus is commonly used for keeping data organized and 
displaying it.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># generate an example character matrix as a pandas dataframe</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;alpha&quot;</span><span class="p">,</span> <span class="s2">&quot;beta&quot;</span><span class="p">,</span> <span class="s2">&quot;gamma&quot;</span><span class="p">,</span> <span class="s2">&quot;delta&quot;</span><span class="p">,</span> <span class="s2">&quot;epsilon&quot;</span><span class="p">],</span>
    <span class="n">data</span><span class="o">=</span><span class="n">arr</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># display the dataframe</span>
<span class="n">data</span>
</code></pre></div>
<figure>
<table>
<thead>
<tr>
<th align="left"></th>
<th align="right">0</th>
<th align="right">1</th>
<th align="right">2</th>
<th align="right">3</th>
<th align="right">4</th>
<th align="right">5</th>
<th align="right">6</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">alpha</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr>
<td align="left">beta</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr>
<td align="left">gamma</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr>
<td align="left">delta</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr>
<td align="left">epsilon</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
</figure>
<p><mark>add plot of data on the tips of tree here, demonstrating indexing from DF.</mark></p>
<h2 id="array-operations">Array Operations<a class="headerlink" href="#array-operations" title="Permanent link">&para;</a></h2>
<p>Although dataframes are great for clarity and organization, they are 
considerably slower than working with <code>numpy</code> arrays, and so when writing
efficient code I tend to extract the array from a dataframe and operate on
it directly. The example below is a somewhat complex function, named 
<code>preprocess_matrix</code>, that we will use in this lesson to greatly speed up 
calculations of parsimony scores from a character matrix. </p>
<p>The goal of this function is to reduce the total number of characters that 
actually need to be analyzed independently. In this respect, it performs
three operations: (1) we can skip any data columns that are invariant, such 
as "t6" in the table above, since we know that these will have zero events 
under parsimony; (2) Similarly, we can skip any columns that have only a 
single difference, such as "t5", since we know that regardless of the tree 
topology this will require only a 1 event under parsimony, and so we can 
simply count 1 for each of these; (3) Finally, we can reduce the amount of 
work by counting how many unique site patterns there are, and only applying 
our algorithm to these, and again multiplying the result by the number of 
times it occured. For example, columns "t3" and "t4" are identical, and 
similarly, "t0" and "t2" are the same, but simply inverted. </p>
<p>Tricks like this are common in phylogenetic algorithms, especially
when dealing with very large character matrices. Reducing the number of 
unique characters that need to be independently analyzed will have a large
effect when the algorithm must be repeated thousands of times, as is common
when performing a full tree inference algorithm. The full <code>preprocess_matrix</code>
function can be found below. We will go through it step by step to describe 
each operation, with reference to the line numbers of code blocks.</p>
<details class="example">
<summary>preprocess_matrix function definition</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Tuple</span>

<span class="k">def</span> <span class="nf">preprocess_matrix</span><span class="p">(</span><span class="n">data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return a score and dict of sites to be processed.</span>

<span class="sd">    This func will remove invariant sites and sites with only a single</span>
<span class="sd">    autapomorphy to get a reduced DataFrame with only unique patterns. </span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data: pandas.DataFrame</span>
<span class="sd">        A pandas dataframe character matrix with taxon names as index.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Tuple </span>
<span class="sd">        A 3-part tuple with (score, counts, data). The score is an int </span>
<span class="sd">        for the number of singleton sites; the counts is an array of ints</span>
<span class="sd">        for the number of occurrence of each unique site pattern; the data</span>
<span class="sd">        is a DataFrame with the unique site patterns.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># get data as a numpy array</span>
    <span class="n">arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

    <span class="c1"># get int array counting how many samples match the state of </span>
    <span class="c1"># the first sample in each column. We will use this for filtering.</span>
    <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">arr</span> <span class="o">==</span> <span class="n">arr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># get boolean array of invariant columns, i.e., all items match the first.</span>
    <span class="n">invariant</span> <span class="o">=</span> <span class="n">counts</span> <span class="o">==</span> <span class="n">arr</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># get boolean array of singleton columns, i.e., nmaching is 1 or N-1</span>
    <span class="n">singletons</span> <span class="o">=</span> <span class="p">(</span><span class="n">counts</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">counts</span> <span class="o">==</span> <span class="n">arr</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>

    <span class="c1"># combine boolean masks and filter columns from data</span>
    <span class="n">arr</span> <span class="o">=</span> <span class="n">arr</span><span class="p">[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">invert</span><span class="p">(</span><span class="n">invariant</span> <span class="o">|</span> <span class="n">singletons</span><span class="p">)]</span>

    <span class="c1"># count the number of singleton sites</span>
    <span class="n">score</span> <span class="o">=</span> <span class="n">singletons</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

    <span class="c1"># get unique site patterns and their counts</span>
    <span class="n">arr</span><span class="p">,</span> <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">arr</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># find columns that are the same PATTERN as others, i.e., recode patterns</span>
    <span class="c1"># such that [0, 1, 1, 2] == [1, 2, 2, 0] == [2, 0, 0, 1], etc.</span>
    <span class="n">uniq_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">cidx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">arr</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>

        <span class="c1"># get a column of data. e.g., [1, 1, 0, 1, 2]</span>
        <span class="n">col</span> <span class="o">=</span> <span class="n">arr</span><span class="p">[:,</span> <span class="n">cidx</span><span class="p">]</span>

        <span class="c1"># get index of first occurrence of each unique base, e.g., [0, 2, 4],</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">idxs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">col</span><span class="p">,</span> <span class="n">return_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># get order of indices, e.g., [0, 1, 2]</span>
        <span class="n">order</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">idxs</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="c1"># get site as ordered index of unique patterns, e.g., [0, 0, 1, 0, 2]</span>
        <span class="n">site</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">((</span><span class="n">order</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">col</span><span class="p">))</span>

        <span class="c1"># record each uniq pattern, or add to it if already mirrored.</span>
        <span class="k">if</span> <span class="n">site</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">uniq_dict</span><span class="p">:</span>
            <span class="n">uniq_dict</span><span class="p">[</span><span class="n">site</span><span class="p">]</span> <span class="o">=</span> <span class="n">counts</span><span class="p">[</span><span class="n">cidx</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">uniq_dict</span><span class="p">[</span><span class="n">site</span><span class="p">]</span> <span class="o">+=</span> <span class="n">counts</span><span class="p">[</span><span class="n">cidx</span><span class="p">]</span>

    <span class="c1"># convert unique patterns and counts back into a dataframe and array</span>
    <span class="n">arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">column_stack</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">uniq_dict</span><span class="p">))</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">arr</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">uniq_dict</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

    <span class="c1"># return the score, counts, and data</span>
    <span class="k">return</span> <span class="n">score</span><span class="p">,</span> <span class="n">counts</span><span class="p">,</span> <span class="n">data</span>
</code></pre></div></td></tr></table></div>
</details>
<p>As an example, when we apply this function to our data it is reduced from 
7 sites to only 3 unique sites, two of which occurred twice, and one that
occurred once. In addition, the parsimony score for this matrix is already
1, since there was one site that had a pattern with only one sample different
from the others. </p>
<h4 id="pandas-vs-numpy-speed">Pandas vs Numpy speed<a class="headerlink" href="#pandas-vs-numpy-speed" title="Permanent link">&para;</a></h4>
<p>This preprocessing function will only need to be run once, at the beginning
of a tree inference program, and therefore its overall speed is not of 
great concern. However, in many other functions that we write the speed of
the code will be important, since the function may need to be iterated many
thousands of times, such as a function to calculate the fitch parsimony 
score. Thus it is useful to learn some tricks for writing more efficient 
Python code.</p>
<p>One trick that I mentioned concerns the use of <code>pandas</code> versus <code>numpy</code>. A
quick demonstration can show clearly that <code>numpy</code> is much faster, even for
very simple operations like 2-dimensional indexing. Let's look a quick example, 
using our handy <code>%%timeit</code> magic command. </p>
<div class="highlight"><pre><span></span><code><span class="c1"># create a numpy array of all zeros of size (100 x 100)</span>
<span class="n">arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">100</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>

<span class="c1"># create a dataframe containing the same data</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">arr</span><span class="p">)</span>
</code></pre></div>
<p>First let's measure the time to index a single element using <code>numpy</code>:
<div class="highlight"><pre><span></span><code><span class="o">%%</span><span class="n">timeit</span>
<span class="n">arr</span><span class="p">[</span><span class="mi">50</span><span class="p">,</span> <span class="mi">50</span><span class="p">]</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="c1"># 90.7 ns ± 2.88 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)</span>
</code></pre></div>
And next let's index a single element using <code>pandas</code>:
<div class="highlight"><pre><span></span><code><span class="o">%%</span><span class="n">timeit</span>
<span class="n">data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">50</span><span class="p">,</span> <span class="mi">50</span><span class="p">]</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="c1"># 20.1 µs ± 447 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)</span>
</code></pre></div>
Indexing with <code>numpy</code> is <em>&gt;200X faster!</em> Thus, if your function involves 
indexing or performing almost any operation over a dataframe many times, it
almost always beneficial to convert it to a numpy array first. </p>
<p>For example, in our preprocessing function, we convert the dataframe to an 
array at the beginning on line 23...</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">22</span>
<span class="normal">23</span></pre></div></td><td class="code"><div><pre><span></span><code>    <span class="c1"># get data as a numpy array</span>
    <span class="n">arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<p>... then we perform operations of the array, before finally converting it
back to a dataframe at the end, on line 69, so that the data is returned 
to the user still containing nice labels.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">69</span></pre></div></td><td class="code"><div><pre><span></span><code>    <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">arr</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<h4 id="boolean-masking">Boolean masking<a class="headerlink" href="#boolean-masking" title="Permanent link">&para;</a></h4>
<p>One of the fastest and most useful ways to filter data in <code>numpy</code> arrays
is by using <em>boolean masking</em>. This involves generating an array full
of <code>boolean</code> values and using it to index the original array. This will 
return the values of the array at all positions where the boolean
mask is <code class="highlight"><span class="kc">True</span></code>, and not return where it is <code class="highlight"><span class="kc">False</span></code>. This
can be applied to individual cells of an array, but even more useful is
to apply it to columns or rows of an array. Let's walk through the 
example below:</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span></pre></div></td><td class="code"><div><pre><span></span><code>    <span class="c1"># get int array counting how many samples match the state of </span>
    <span class="c1"># the first sample in each column. We will use this for filtering.</span>
    <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">arr</span> <span class="o">==</span> <span class="n">arr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
<p>This one line is actually doing something fairly complex, so let's look in
more detail. Below you can see the value of the original <code>arr</code>, as well 
as of the value represented by the operation <em>inside</em> of the function, 
<code>arr == arr[0]</code>. This statement, which uses <code>==</code> operator, is asking "are
these things equal?". Because this statement is applied to a <code>numpy</code> array, 
it is <em>broadcast</em> to apply to each item. In this case, that involves asking 
for each item, whether it matches the corresponding item in the first row. </p>
<p>A bit tricky, right? But this turns out to be a very useful and common 
operation when working with arrays of this type. The result of this statement
is shown below as a variable named <code>mask</code>. You can see that it is a boolean 
array where the first row is all True, but other items may be True or False.
This function will work regardless of whether the data is binary or composed
of some other multistate characters.</p>
<p>The final statement below is the same as in our function. It takes the 
boolean array (which we named <code>mask</code>) and calculates the sum over <code>axis=0</code>, 
which means to get the sum of each column. Our final result is an integer 
array with the number of values matching the first row item. As we'll see, 
these counts are useful for applying our next filters to ask whether a site
is invariant, or if it contains only one difference.</p>
<p><div class="highlight"><pre><span></span><code><span class="c1"># view the original data arr</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;arr:</span><span class="se">\n</span><span class="si">{</span><span class="n">arr</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># compare arr columns-wise to the value in the first row of arr</span>
<span class="n">mask</span> <span class="o">=</span> <span class="n">arr</span> <span class="o">==</span> <span class="n">arr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">mask:</span><span class="se">\n</span><span class="si">{</span><span class="n">mask</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># get sum of True values in each column of the boolean mask</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">arr</span> <span class="o">==</span> <span class="n">arr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">counts:</span><span class="se">\n</span><span class="si">{</span><span class="n">counts</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="c1"># arr:</span>
<span class="c1"># [[1 0 0 1 1 0 0]</span>
<span class="c1">#  [0 0 1 0 0 0 0]</span>
<span class="c1">#  [1 1 0 0 0 0 0]</span>
<span class="c1">#  [1 1 0 1 1 1 0]</span>
<span class="c1">#  [0 0 1 1 1 0 0]]</span>

<span class="c1"># mask:</span>
<span class="c1"># [[ True  True  True  True  True  True  True]</span>
<span class="c1">#  [False  True False False False  True  True]</span>
<span class="c1">#  [ True False  True False False  True  True]</span>
<span class="c1">#  [ True False  True  True  True False  True]</span>
<span class="c1">#  [False  True False  True  True  True  True]]</span>

<span class="c1"># counts:</span>
<span class="c1"># [3 3 3 3 3 4 5]</span>
</code></pre></div></p>
<p><mark>TODO</mark></p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span></pre></div></td><td class="code"><div><pre><span></span><code>    <span class="c1"># get boolean array of invariant columns, i.e., all items match the first.</span>
    <span class="n">invariant</span> <span class="o">=</span> <span class="n">counts</span> <span class="o">==</span> <span class="n">arr</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># get boolean array of singleton columns, i.e., nmaching is 1 or N-1</span>
    <span class="n">singletons</span> <span class="o">=</span> <span class="p">(</span><span class="n">counts</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">counts</span> <span class="o">==</span> <span class="n">arr</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>

    <span class="c1"># combine boolean masks and filter columns from data</span>
    <span class="n">arr</span> <span class="o">=</span> <span class="n">arr</span><span class="p">[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">invert</span><span class="p">(</span><span class="n">invariant</span> <span class="o">|</span> <span class="n">singletons</span><span class="p">)]</span>
</code></pre></div></td></tr></table></div>
<h4 id="array-indexing">Array indexing<a class="headerlink" href="#array-indexing" title="Permanent link">&para;</a></h4>
<p><mark>TODO</mark></p>
<h4 id="unique-patterns">Unique patterns<a class="headerlink" href="#unique-patterns" title="Permanent link">&para;</a></h4>
<p><mark>TODO</mark></p>
<div class="highlight"><pre><span></span><code><span class="c1"># example to reduce character matrix to unique patterns</span>
<span class="n">score</span><span class="p">,</span> <span class="n">counts</span><span class="p">,</span> <span class="n">reduced_data</span> <span class="o">=</span> <span class="n">preprocess_matrix</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">reduced_data</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="c1"># 1</span>
<span class="c1"># [1 2 2]</span>
<span class="c1">#          0  1  2</span>
<span class="c1"># alpha    0  0  0</span>
<span class="c1"># beta     0  1  1</span>
<span class="c1"># gamma    1  0  1</span>
<span class="c1"># delta    1  0  0</span>
<span class="c1"># epsilon  0  1  0</span>
</code></pre></div>
<h2 id="fitch-parsimony-function">Fitch Parsimony Function<a class="headerlink" href="#fitch-parsimony-function" title="Permanent link">&para;</a></h2>
<p>Now we can compute the rest of the parsimony score by 
analyzing these remaining unique patterns on the tree. For this, we can 
use a modification of our function from the last page which includes our
<code>preprocess_matrix</code> function within it.</p>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">fitch_parsimony</span><span class="p">(</span><span class="n">tree</span><span class="p">:</span> <span class="n">toytree</span><span class="o">.</span><span class="n">ToyTree</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return Fitch parsimony score given a tree and single trait.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    tree: ToyTree</span>
<span class="sd">        A tree on which to count state changes.</span>
<span class="sd">    data: DataFrame</span>
<span class="sd">        A DataFrame with tree tip labels in the index and &gt;=1 columns</span>
<span class="sd">        of discrete character data as int or str.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        The minimum changes required for trait data to evolve on this tree.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># reduce matrix to unique sites and counts, and get initial score</span>
    <span class="n">score</span><span class="p">,</span> <span class="n">counts</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">preprocess_matrix</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

    <span class="c1"># sort data to tip idxorder and convert to array (for faster indexing)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">tree</span><span class="o">.</span><span class="n">get_tip_labels</span><span class="p">()]</span><span class="o">.</span><span class="n">values</span>

    <span class="c1"># iterate over each column index of data</span>
    <span class="k">for</span> <span class="n">cidx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>

        <span class="c1"># iterate over Nodes in idxorder (postorder sorted) traversal</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree</span><span class="p">:</span>

            <span class="c1"># leaves are visited first, and converted to a set type</span>
            <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">():</span>
                <span class="n">char</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">idx</span><span class="p">,</span> <span class="n">cidx</span><span class="p">]</span>
                <span class="n">node</span><span class="o">.</span><span class="n">_fitch</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">char</span><span class="p">])</span>

            <span class="c1"># internal Nodes examine the sets of their children&#39;s states</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">node</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">():</span>

                <span class="c1"># check for shared (intersecting) states</span>
                <span class="n">shared</span> <span class="o">=</span> <span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">_fitch</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="p">))</span>

                <span class="c1"># if any states are shared then ancestor inherits this state</span>
                <span class="k">if</span> <span class="n">shared</span><span class="p">:</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">_fitch</span> <span class="o">=</span> <span class="n">shared</span>

                <span class="c1"># if none shared, then store the union and increment counter</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">_fitch</span> <span class="o">=</span> <span class="nb">set</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="o">*</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">_fitch</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="p">))</span>
                    <span class="n">score</span> <span class="o">+=</span> <span class="n">counts</span><span class="p">[</span><span class="n">cidx</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">score</span>
</code></pre></div>
<p>An implementation on our example dataset:
<div class="highlight"><pre><span></span><code><span class="n">fitch_parsimony</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
</code></pre></div></p>
<div class="highlight"><pre><span></span><code><span class="c1"># 5</span>
</code></pre></div>
<!-- 
If we wanted to calculate additional information about our character 
matrix we might return the parsimony score for each trait individually, 
perhaps as an array, instead of just returning the sum, so that we could
identify which characters are good or poor fits to a particular tree.

...
-->


  




                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.instant", "navigation.tracking", "search.highlight", "search.suggest", "content.code.annotate"], "search": "../../assets/javascripts/workers/search.db81ec45.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.6df46069.min.js"></script>
      
        <script src="../../javascripts/extra.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>